{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "plan: \n",
    "with script, make a string of the following format that includes all haplotypes and also filters on both cwd version 2 and 3 (must be 'C' (common) for both):\n",
    "\n",
    "https://www.ebi.ac.uk/ipd/imgt/hla/alleles/?query=or(and(eq(cwd.version_3.Total,%22C%22),eq(cwd.version_2,%22C%22),startsWith(name,%22A*24:02%22)),and(eq(cwd.version_3.Total,%22C%22),eq(cwd.version_2,%22C%22),startsWith(name,%22A*01:02%22)))\n",
    "\n",
    "paste resulting url string into browser and then download the result list mapping IPD Accession (what we need for getting jsons) to allele name.\n",
    "then loop over all the accessions to get json urls: https://www.ebi.ac.uk/cgi-bin/ipd/api/allele/[HLA00002]\n",
    "    format: https://www.ebi.ac.uk/cgi-bin/ipd/api/allele/[accession]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import json\n",
    "from Bio.Seq import Seq\n",
    "import math\n",
    "import urllib.request"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def readHLA(filename):\n",
    "    \"\"\"\n",
    "    takes in filename of HLA locus xlsx, outputs list of most frequent alleles\n",
    "    an allele is defined as frequent if it occurs in more than 1% of the population for any ethnic group \n",
    "    \"\"\"\n",
    "    df = pd.read_excel(filename)\n",
    "    freq_alleles_list = []\n",
    "    cutoff = 0.01\n",
    "    for index, row in df.iterrows():\n",
    "        # print(row[\"A\"])\n",
    "        if max(row[1], row[3], row[5], row[7], row[9], row[11], row[13], row[15], row[17], row[19], row[21], row[23], row[25], row[27], row[29], row[31], row[33], row[35], row[37], row[39], row[41], row[43], row[45], row[47], row[49], row[51]) >= cutoff:\n",
    "            # freq_alleles_list.append(row[0])\n",
    "            if row[0][-1] == \"g\":\n",
    "                freq_alleles_list.append(row[0][:-1])\n",
    "            else:\n",
    "                freq_alleles_list.append(row[0])\n",
    "    return freq_alleles_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/sm/2bd25dyd1bx5nl6h9b6j4z7c0000gn/T/ipykernel_14161/386535371.py:11: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  if max(row[1], row[3], row[5], row[7], row[9], row[11], row[13], row[15], row[17], row[19], row[21], row[23], row[25], row[27], row[29], row[31], row[33], row[35], row[37], row[39], row[41], row[43], row[45], row[47], row[49], row[51]) >= cutoff:\n",
      "/var/folders/sm/2bd25dyd1bx5nl6h9b6j4z7c0000gn/T/ipykernel_14161/386535371.py:13: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  if row[0][-1] == \"g\":\n",
      "/var/folders/sm/2bd25dyd1bx5nl6h9b6j4z7c0000gn/T/ipykernel_14161/386535371.py:16: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  freq_alleles_list.append(row[0])\n",
      "/var/folders/sm/2bd25dyd1bx5nl6h9b6j4z7c0000gn/T/ipykernel_14161/386535371.py:14: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  freq_alleles_list.append(row[0][:-1])\n",
      "/var/folders/sm/2bd25dyd1bx5nl6h9b6j4z7c0000gn/T/ipykernel_14161/386535371.py:11: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  if max(row[1], row[3], row[5], row[7], row[9], row[11], row[13], row[15], row[17], row[19], row[21], row[23], row[25], row[27], row[29], row[31], row[33], row[35], row[37], row[39], row[41], row[43], row[45], row[47], row[49], row[51]) >= cutoff:\n",
      "/var/folders/sm/2bd25dyd1bx5nl6h9b6j4z7c0000gn/T/ipykernel_14161/386535371.py:13: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  if row[0][-1] == \"g\":\n",
      "/var/folders/sm/2bd25dyd1bx5nl6h9b6j4z7c0000gn/T/ipykernel_14161/386535371.py:14: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  freq_alleles_list.append(row[0][:-1])\n",
      "/var/folders/sm/2bd25dyd1bx5nl6h9b6j4z7c0000gn/T/ipykernel_14161/386535371.py:16: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  freq_alleles_list.append(row[0])\n",
      "/var/folders/sm/2bd25dyd1bx5nl6h9b6j4z7c0000gn/T/ipykernel_14161/386535371.py:11: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  if max(row[1], row[3], row[5], row[7], row[9], row[11], row[13], row[15], row[17], row[19], row[21], row[23], row[25], row[27], row[29], row[31], row[33], row[35], row[37], row[39], row[41], row[43], row[45], row[47], row[49], row[51]) >= cutoff:\n",
      "/var/folders/sm/2bd25dyd1bx5nl6h9b6j4z7c0000gn/T/ipykernel_14161/386535371.py:13: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  if row[0][-1] == \"g\":\n",
      "/var/folders/sm/2bd25dyd1bx5nl6h9b6j4z7c0000gn/T/ipykernel_14161/386535371.py:14: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  freq_alleles_list.append(row[0][:-1])\n",
      "/var/folders/sm/2bd25dyd1bx5nl6h9b6j4z7c0000gn/T/ipykernel_14161/386535371.py:16: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  freq_alleles_list.append(row[0])\n",
      "/var/folders/sm/2bd25dyd1bx5nl6h9b6j4z7c0000gn/T/ipykernel_14161/386535371.py:11: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  if max(row[1], row[3], row[5], row[7], row[9], row[11], row[13], row[15], row[17], row[19], row[21], row[23], row[25], row[27], row[29], row[31], row[33], row[35], row[37], row[39], row[41], row[43], row[45], row[47], row[49], row[51]) >= cutoff:\n",
      "/var/folders/sm/2bd25dyd1bx5nl6h9b6j4z7c0000gn/T/ipykernel_14161/386535371.py:13: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  if row[0][-1] == \"g\":\n",
      "/var/folders/sm/2bd25dyd1bx5nl6h9b6j4z7c0000gn/T/ipykernel_14161/386535371.py:14: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  freq_alleles_list.append(row[0][:-1])\n",
      "/var/folders/sm/2bd25dyd1bx5nl6h9b6j4z7c0000gn/T/ipykernel_14161/386535371.py:16: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  freq_alleles_list.append(row[0])\n",
      "/var/folders/sm/2bd25dyd1bx5nl6h9b6j4z7c0000gn/T/ipykernel_14161/386535371.py:11: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  if max(row[1], row[3], row[5], row[7], row[9], row[11], row[13], row[15], row[17], row[19], row[21], row[23], row[25], row[27], row[29], row[31], row[33], row[35], row[37], row[39], row[41], row[43], row[45], row[47], row[49], row[51]) >= cutoff:\n",
      "/var/folders/sm/2bd25dyd1bx5nl6h9b6j4z7c0000gn/T/ipykernel_14161/386535371.py:13: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  if row[0][-1] == \"g\":\n",
      "/var/folders/sm/2bd25dyd1bx5nl6h9b6j4z7c0000gn/T/ipykernel_14161/386535371.py:16: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  freq_alleles_list.append(row[0])\n",
      "/var/folders/sm/2bd25dyd1bx5nl6h9b6j4z7c0000gn/T/ipykernel_14161/386535371.py:14: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  freq_alleles_list.append(row[0][:-1])\n"
     ]
    }
   ],
   "source": [
    "freq_alleles = {}\n",
    "freq_alleles[\"A\"] = readHLA(\"./HLA_allele_freq_sheets/A.xlsx\")\n",
    "freq_alleles[\"B\"] = readHLA(\"./HLA_allele_freq_sheets/B.xlsx\")\n",
    "freq_alleles[\"C\"] = readHLA(\"./HLA_allele_freq_sheets/C.xlsx\")\n",
    "freq_alleles[\"DQB1\"] = readHLA(\"./HLA_allele_freq_sheets/DQB1.xlsx\")\n",
    "freq_alleles[\"DRB1\"] = readHLA(\"./HLA_allele_freq_sheets/DRB1.xlsx\")\n",
    "# freq_alleles is a dict mapping strings of gene (eg. \"A\", \"B\") to list of common haplotype names (strings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'A': ['A*30:04',\n",
       "  'A*24:03',\n",
       "  'A*24:02',\n",
       "  'A*11:01',\n",
       "  'A*34:02',\n",
       "  'A*32:01',\n",
       "  'A*01:01',\n",
       "  'A*03:02',\n",
       "  'A*68:01',\n",
       "  'A*23:01',\n",
       "  'A*11:02',\n",
       "  'A*74:01',\n",
       "  'A*02:05',\n",
       "  'A*26:01',\n",
       "  'A*30:01',\n",
       "  'A*34:01',\n",
       "  'A*26:03',\n",
       "  'A*66:01',\n",
       "  'A*03:01',\n",
       "  'A*02:11',\n",
       "  'A*02:02',\n",
       "  'A*02:03',\n",
       "  'A*26:02',\n",
       "  'A*33:01',\n",
       "  'A*30:02',\n",
       "  'A*24:07',\n",
       "  'A*29:02',\n",
       "  'A*25:01',\n",
       "  'A*68:02',\n",
       "  'A*33:03',\n",
       "  'A*36:01',\n",
       "  'A*29:01',\n",
       "  'A*02:01',\n",
       "  'A*02:17',\n",
       "  'A*31:01',\n",
       "  'A*68:03',\n",
       "  'A*02:06',\n",
       "  'A*02:07'],\n",
       " 'B': ['B*54:01',\n",
       "  'B*55:01',\n",
       "  'B*59:01',\n",
       "  'B*44:02',\n",
       "  'B*81:01',\n",
       "  'B*13:02',\n",
       "  'B*07:02',\n",
       "  'B*07:05',\n",
       "  'B*50:01',\n",
       "  'B*35:12',\n",
       "  'B*51:06',\n",
       "  'B*13:01',\n",
       "  'B*41:01',\n",
       "  'B*15:11',\n",
       "  'B*39:06',\n",
       "  'B*57:01',\n",
       "  'B*44:03',\n",
       "  'B*51:02',\n",
       "  'B*15:35',\n",
       "  'B*56:01',\n",
       "  'B*15:16',\n",
       "  'B*57:03',\n",
       "  'B*27:04',\n",
       "  'B*40:10',\n",
       "  'B*35:01',\n",
       "  'B*27:05',\n",
       "  'B*58:02',\n",
       "  'B*58:01',\n",
       "  'B*42:01',\n",
       "  'B*15:01',\n",
       "  'B*15:15',\n",
       "  'B*15:06',\n",
       "  'B*39:10',\n",
       "  'B*15:03',\n",
       "  'B*14:02',\n",
       "  'B*49:01',\n",
       "  'B*35:02',\n",
       "  'B*51:01',\n",
       "  'B*46:01',\n",
       "  'B*35:08',\n",
       "  'B*15:21',\n",
       "  'B*55:02',\n",
       "  'B*40:02',\n",
       "  'B*15:13',\n",
       "  'B*35:17',\n",
       "  'B*53:01',\n",
       "  'B*15:10',\n",
       "  'B*38:01',\n",
       "  'B*56:02',\n",
       "  'B*15:12',\n",
       "  'B*56:04',\n",
       "  'B*37:01',\n",
       "  'B*35:43',\n",
       "  'B*27:06',\n",
       "  'B*35:03',\n",
       "  'B*18:01',\n",
       "  'B*48:01',\n",
       "  'B*35:05',\n",
       "  'B*15:17',\n",
       "  'B*40:06',\n",
       "  'B*52:01',\n",
       "  'B*15:27',\n",
       "  'B*15:18',\n",
       "  'B*14:01',\n",
       "  'B*38:02',\n",
       "  'B*39:01',\n",
       "  'B*15:25',\n",
       "  'B*39:05',\n",
       "  'B*08:01',\n",
       "  'B*45:01',\n",
       "  'B*67:01',\n",
       "  'B*15:02',\n",
       "  'B*40:01',\n",
       "  'B*15:07'],\n",
       " 'C': ['C*07:02',\n",
       "  'C*14:02',\n",
       "  'C*07:04',\n",
       "  'C*08:04',\n",
       "  'C*08:01',\n",
       "  'C*15:02',\n",
       "  'C*03:04',\n",
       "  'C*08:06',\n",
       "  'C*12:04',\n",
       "  'C*03:02',\n",
       "  'C*17:01',\n",
       "  'C*08:03',\n",
       "  'C*04:01',\n",
       "  'C*07:01',\n",
       "  'C*16:01',\n",
       "  'C*06:02',\n",
       "  'C*05:01',\n",
       "  'C*03:03',\n",
       "  'C*12:02',\n",
       "  'C*03:06',\n",
       "  'C*16:02',\n",
       "  'C*01:02',\n",
       "  'C*03:05',\n",
       "  'C*14:03',\n",
       "  'C*02:02',\n",
       "  'C*16:04',\n",
       "  'C*18:01',\n",
       "  'C*04:06',\n",
       "  'C*04:03',\n",
       "  'C*15:05',\n",
       "  'C*12:03',\n",
       "  'C*08:02'],\n",
       " 'DQB1': ['DQB1*03:02',\n",
       "  'DQB1*02:01',\n",
       "  'DQB1*05:02',\n",
       "  'DQB1*03:01',\n",
       "  'DQB1*05:03',\n",
       "  'DQB1*06:04',\n",
       "  'DQB1*04:02',\n",
       "  'DQB1*06:09',\n",
       "  'DQB1*06:02',\n",
       "  'DQB1*03:03',\n",
       "  'DQB1*04:01',\n",
       "  'DQB1*06:01',\n",
       "  'DQB1*05:01',\n",
       "  'DQB1*06:03'],\n",
       " 'DRB1': ['DRB1*14:05',\n",
       "  'DRB1*14:04',\n",
       "  'DRB1*15:02',\n",
       "  'DRB1*04:05',\n",
       "  'DRB1*08:02',\n",
       "  'DRB1*04:07',\n",
       "  'DRB1*15:06',\n",
       "  'DRB1*13:02',\n",
       "  'DRB1*09:01',\n",
       "  'DRB1*08:04',\n",
       "  'DRB1*11:04',\n",
       "  'DRB1*14:08',\n",
       "  'DRB1*04:06',\n",
       "  'DRB1*15:03',\n",
       "  'DRB1*13:03',\n",
       "  'DRB1*04:02',\n",
       "  'DRB1*12:01',\n",
       "  'DRB1*04:01',\n",
       "  'DRB1*01:02',\n",
       "  'DRB1*11:01',\n",
       "  'DRB1*14:02',\n",
       "  'DRB1*15:01',\n",
       "  'DRB1*03:02',\n",
       "  'DRB1*14:01',\n",
       "  'DRB1*03:01',\n",
       "  'DRB1*01:01',\n",
       "  'DRB1*12:02',\n",
       "  'DRB1*14:03',\n",
       "  'DRB1*13:04',\n",
       "  'DRB1*08:01',\n",
       "  'DRB1*04:04',\n",
       "  'DRB1*14:06',\n",
       "  'DRB1*13:01',\n",
       "  'DRB1*13:12',\n",
       "  'DRB1*11:02',\n",
       "  'DRB1*01:03',\n",
       "  'DRB1*16:01',\n",
       "  'DRB1*04:11',\n",
       "  'DRB1*04:10',\n",
       "  'DRB1*07:01',\n",
       "  'DRB1*10:01',\n",
       "  'DRB1*04:03',\n",
       "  'DRB1*16:02',\n",
       "  'DRB1*08:03']}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "freq_alleles"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### HLA-A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "### overfilters \n",
    "# prefix = \"https://www.ebi.ac.uk/ipd/imgt/hla/alleles/?query=or(\"\n",
    "# a_str = str()\n",
    "# for allele_name in freq_alleles[\"A\"]:\n",
    "#     #print(allele_name)\n",
    "#     s = f\",and(eq(cwd.version_3.Total,%22C%22),eq(cwd.version_2,%22C%22),startsWith(name,%22{allele_name}%22))\"\n",
    "#     a_str+=s\n",
    "# a_str = a_str[1:]\n",
    "\n",
    "# a_str = prefix+a_str+\")\"\n",
    "# a_str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'https://www.ebi.ac.uk/ipd/imgt/hla/alleles/?query=or(and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*30:04%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*24:03%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*24:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*11:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*34:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*32:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*01:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*03:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*68:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*23:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*11:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*74:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*02:05%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*26:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*30:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*34:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*26:03%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*66:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*03:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*02:11%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*02:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*02:03%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*26:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*33:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*30:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*24:07%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*29:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*25:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*68:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*33:03%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*36:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*29:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*02:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*02:17%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*31:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*68:03%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*02:06%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22A*02:07%22)))'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## only using CWD version 2 to filter seems to work well \n",
    "prefix = \"https://www.ebi.ac.uk/ipd/imgt/hla/alleles/?query=or(\"\n",
    "a_str = str()\n",
    "for allele_name in freq_alleles[\"A\"]:\n",
    "    #print(allele_name)\n",
    "    s = f\",and(eq(cwd.version_2,%22C%22),startsWith(name,%22{allele_name}%22))\"\n",
    "    a_str+=s\n",
    "a_str = a_str[1:]\n",
    "\n",
    "a_str = prefix+a_str+\")\"\n",
    "a_str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'A*30:04:01:01': 'HLA00092',\n",
       " 'A*24:03:01:01': 'HLA00053',\n",
       " 'A*24:02:01:01': 'HLA00050',\n",
       " 'A*11:01:01:01': 'HLA00043',\n",
       " 'A*34:02:01:01': 'HLA00109',\n",
       " 'A*32:01:01:01': 'HLA00101',\n",
       " 'A*01:01:01:01': 'HLA00001',\n",
       " 'A*03:02:01:01': 'HLA00040',\n",
       " 'A*68:01:01:01': 'HLA00115',\n",
       " 'A*68:01:02:01': 'HLA00116',\n",
       " 'A*23:01:01:01': 'HLA00048',\n",
       " 'A*11:02:01:01': 'HLA00044',\n",
       " 'A*74:01:01:01': 'HLA00127',\n",
       " 'A*02:05:01:01': 'HLA00010',\n",
       " 'A*26:01:01:01': 'HLA00073',\n",
       " 'A*30:01:01:01': 'HLA00089',\n",
       " 'A*34:01:01:01': 'HLA00108',\n",
       " 'A*26:03:01:01': 'HLA00075',\n",
       " 'A*66:01:01:01': 'HLA00112',\n",
       " 'A*03:01:01:01': 'HLA00037',\n",
       " 'A*02:11:01:01': 'HLA00016',\n",
       " 'A*02:02:01:01': 'HLA00007',\n",
       " 'A*02:03:01:01': 'HLA00008',\n",
       " 'A*33:01:01:01': 'HLA00104',\n",
       " 'A*30:02:01:01': 'HLA00090',\n",
       " 'A*24:07:01:01': 'HLA00057',\n",
       " 'A*29:02:01:01': 'HLA00086',\n",
       " 'A*25:01:01:01': 'HLA00071',\n",
       " 'A*68:02:01:01': 'HLA00117',\n",
       " 'A*33:03:01:01': 'HLA00106',\n",
       " 'A*36:01:01:01': 'HLA00110',\n",
       " 'A*29:01:01:01': 'HLA00085',\n",
       " 'A*02:01:01:01': 'HLA00005',\n",
       " 'A*02:01:04': 'HLA01032',\n",
       " 'A*31:01:02:01': 'HLA00097',\n",
       " 'A*68:03:01:01': 'HLA00118',\n",
       " 'A*02:06:01:01': 'HLA00011',\n",
       " 'A*02:07:01:01': 'HLA00012',\n",
       " 'A*26:02:01': 'HLA00074',\n",
       " 'A*02:17:02:01': 'HLA00023'}"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_to_a = pd.read_csv(\"HLA_name_to_dbAccession/HLA_A_name_to_accession.txt\", sep=\"\\t\")\n",
    "n_to_a \n",
    "name_to_accession = dict(zip(n_to_a['name'], n_to_a['accession']))\n",
    "name_to_accession"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'HLA_A_name_to_accession.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mFileNotFoundError\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[25]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m a_out = \u001b[43mpd\u001b[49m\u001b[43m.\u001b[49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mHLA_A_name_to_accession.txt\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[32m      2\u001b[39m a_out\n\u001b[32m      3\u001b[39m a_out[\u001b[33m'\u001b[39m\u001b[33mname\u001b[39m\u001b[33m'\u001b[39m]\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.pyenv/versions/HLAoct2025/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1026\u001b[39m, in \u001b[36mread_csv\u001b[39m\u001b[34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[39m\n\u001b[32m   1013\u001b[39m kwds_defaults = _refine_defaults_read(\n\u001b[32m   1014\u001b[39m     dialect,\n\u001b[32m   1015\u001b[39m     delimiter,\n\u001b[32m   (...)\u001b[39m\u001b[32m   1022\u001b[39m     dtype_backend=dtype_backend,\n\u001b[32m   1023\u001b[39m )\n\u001b[32m   1024\u001b[39m kwds.update(kwds_defaults)\n\u001b[32m-> \u001b[39m\u001b[32m1026\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.pyenv/versions/HLAoct2025/lib/python3.12/site-packages/pandas/io/parsers/readers.py:620\u001b[39m, in \u001b[36m_read\u001b[39m\u001b[34m(filepath_or_buffer, kwds)\u001b[39m\n\u001b[32m    617\u001b[39m _validate_names(kwds.get(\u001b[33m\"\u001b[39m\u001b[33mnames\u001b[39m\u001b[33m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[32m    619\u001b[39m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m620\u001b[39m parser = \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    622\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[32m    623\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.pyenv/versions/HLAoct2025/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1620\u001b[39m, in \u001b[36mTextFileReader.__init__\u001b[39m\u001b[34m(self, f, engine, **kwds)\u001b[39m\n\u001b[32m   1617\u001b[39m     \u001b[38;5;28mself\u001b[39m.options[\u001b[33m\"\u001b[39m\u001b[33mhas_index_names\u001b[39m\u001b[33m\"\u001b[39m] = kwds[\u001b[33m\"\u001b[39m\u001b[33mhas_index_names\u001b[39m\u001b[33m\"\u001b[39m]\n\u001b[32m   1619\u001b[39m \u001b[38;5;28mself\u001b[39m.handles: IOHandles | \u001b[38;5;28;01mNone\u001b[39;00m = \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[32m-> \u001b[39m\u001b[32m1620\u001b[39m \u001b[38;5;28mself\u001b[39m._engine = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.pyenv/versions/HLAoct2025/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1880\u001b[39m, in \u001b[36mTextFileReader._make_engine\u001b[39m\u001b[34m(self, f, engine)\u001b[39m\n\u001b[32m   1878\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[33m\"\u001b[39m\u001b[33mb\u001b[39m\u001b[33m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m mode:\n\u001b[32m   1879\u001b[39m         mode += \u001b[33m\"\u001b[39m\u001b[33mb\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m-> \u001b[39m\u001b[32m1880\u001b[39m \u001b[38;5;28mself\u001b[39m.handles = \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m   1881\u001b[39m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1882\u001b[39m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1883\u001b[39m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mencoding\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1884\u001b[39m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mcompression\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1885\u001b[39m \u001b[43m    \u001b[49m\u001b[43mmemory_map\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mmemory_map\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1886\u001b[39m \u001b[43m    \u001b[49m\u001b[43mis_text\u001b[49m\u001b[43m=\u001b[49m\u001b[43mis_text\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1887\u001b[39m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mencoding_errors\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mstrict\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1888\u001b[39m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mstorage_options\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1889\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1890\u001b[39m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m.handles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[32m   1891\u001b[39m f = \u001b[38;5;28mself\u001b[39m.handles.handle\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.pyenv/versions/HLAoct2025/lib/python3.12/site-packages/pandas/io/common.py:873\u001b[39m, in \u001b[36mget_handle\u001b[39m\u001b[34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[39m\n\u001b[32m    868\u001b[39m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(handle, \u001b[38;5;28mstr\u001b[39m):\n\u001b[32m    869\u001b[39m     \u001b[38;5;66;03m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[32m    870\u001b[39m     \u001b[38;5;66;03m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[32m    871\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m ioargs.encoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[33m\"\u001b[39m\u001b[33mb\u001b[39m\u001b[33m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs.mode:\n\u001b[32m    872\u001b[39m         \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m873\u001b[39m         handle = \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[32m    874\u001b[39m \u001b[43m            \u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    875\u001b[39m \u001b[43m            \u001b[49m\u001b[43mioargs\u001b[49m\u001b[43m.\u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    876\u001b[39m \u001b[43m            \u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m=\u001b[49m\u001b[43mioargs\u001b[49m\u001b[43m.\u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    877\u001b[39m \u001b[43m            \u001b[49m\u001b[43merrors\u001b[49m\u001b[43m=\u001b[49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    878\u001b[39m \u001b[43m            \u001b[49m\u001b[43mnewline\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m    879\u001b[39m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    880\u001b[39m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m    881\u001b[39m         \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n\u001b[32m    882\u001b[39m         handle = \u001b[38;5;28mopen\u001b[39m(handle, ioargs.mode)\n",
      "\u001b[31mFileNotFoundError\u001b[39m: [Errno 2] No such file or directory: 'HLA_A_name_to_accession.txt'"
     ]
    }
   ],
   "source": [
    "a_out = pd.read_csv(\"HLA_A_name_to_accession.txt\", sep=\"\\t\")\n",
    "a_out\n",
    "a_out['name']\n",
    "not_included = []\n",
    "for a_allele in freq_alleles[\"A\"]:\n",
    "    # print(b_allele)\n",
    "    if a_out['name'].str.contains(a_allele).any() == False:\n",
    "        # print(b_allele)\n",
    "        not_included.append(a_allele)\n",
    "\n",
    "not_included"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### HLA-B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "74"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(freq_alleles[\"B\"])\n",
    "# freq_alleles[\"B\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'https://www.ebi.ac.uk/ipd/imgt/hla/alleles/?query=or(and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*54:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*55:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*59:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*44:02%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*81:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*13:02%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*07:02%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*07:05%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*50:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*35:12%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*51:06%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*13:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*41:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*15:11%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*39:06%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*57:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*44:03%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*51:02%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*15:35%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*56:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*15:16%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*57:03%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*27:04%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*40:10%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*35:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*27:05%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*58:02%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*58:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*42:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*15:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*15:15%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*15:06%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*39:10%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*15:03%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*14:02%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*49:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*35:02%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*51:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*46:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*35:08%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*15:21%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*55:02%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*40:02%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*15:13%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*35:17%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*53:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*15:10%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*38:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*56:02%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*15:12%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*56:04%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*37:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*35:43%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*27:06%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*35:03%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*18:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*48:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*35:05%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*15:17%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*40:06%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*52:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*15:27%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*15:18%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*14:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*38:02%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*39:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*15:25%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*39:05%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*08:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*45:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*67:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*15:02%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*40:01%22)),and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22B*15:07%22)))'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## overfilters \n",
    "prefix = \"https://www.ebi.ac.uk/ipd/imgt/hla/alleles/?query=or(\"\n",
    "b_str = str()\n",
    "for allele_name in freq_alleles[\"B\"]:\n",
    "    #print(allele_name)\n",
    "    s = f\",and(eq(cwd.version_3.Total,%22C%22),startsWith(name,%22{allele_name}%22))\"\n",
    "    b_str+=s\n",
    "b_str = b_str[1:]\n",
    "\n",
    "b_str = prefix+b_str+\")\"\n",
    "b_str\n",
    "\n",
    "\n",
    "# ## only using CWD version 2 to filter seems to work well \n",
    "# prefix = \"https://www.ebi.ac.uk/ipd/imgt/hla/alleles/?query=or(\"\n",
    "# b_str = str()\n",
    "# for allele_name in freq_alleles[\"B\"]:\n",
    "#     #print(allele_name)\n",
    "#     s = f\",and(eq(cwd.version_2,%22C%22),startsWith(name,%22{allele_name}%22))\"\n",
    "#     b_str+=s\n",
    "# b_str = b_str[1:]\n",
    "\n",
    "# b_str = prefix+b_str+\")\"\n",
    "# b_str\n",
    "\n",
    "#### PASTE THE RESULTING LINK INTO BROWSER TO GET THE OUTPUTS !! DOWNLOAD FROM BROWSER "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['B*40:10', 'B*15:06', 'B*56:02', 'B*56:04', 'B*15:27']"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# checking which alleles were included from automatic searching and filtering\n",
    "# manually search and get remaining alleles\n",
    "b_out = pd.read_csv(\"HLA_B_rawOut.txt\", sep=\"\\t\")\n",
    "b_out \n",
    "b_out['name']\n",
    "not_included = []\n",
    "for b_allele in freq_alleles[\"B\"]:\n",
    "    # print(b_allele)\n",
    "    if b_out['name'].str.contains(b_allele).any() == False:\n",
    "        # print(b_allele)\n",
    "        not_included.append(b_allele)\n",
    "\n",
    "not_included\n",
    "\n",
    "name_to_accession = dict(zip(b_out['name'], b_out['accession']))\n",
    "name_to_accession"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### HLA-C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "32"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(freq_alleles[\"C\"])\n",
    "# freq_alleles[\"B\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'https://www.ebi.ac.uk/ipd/imgt/hla/alleles/?query=or(and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*07:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*14:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*07:04%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*08:04%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*08:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*15:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*03:04%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*08:06%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*12:04%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*03:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*17:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*08:03%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*04:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*07:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*16:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*06:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*05:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*03:03%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*12:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*03:06%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*16:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*01:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*03:05%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*14:03%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*02:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*16:04%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*18:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*04:06%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*04:03%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*15:05%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*12:03%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22C*08:02%22)))'"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## only using CWD version 2 to filter seems to work well \n",
    "prefix = \"https://www.ebi.ac.uk/ipd/imgt/hla/alleles/?query=or(\"\n",
    "c_str = str()\n",
    "for allele_name in freq_alleles[\"C\"]:\n",
    "    #print(allele_name)\n",
    "    s = f\",and(eq(cwd.version_2,%22C%22),startsWith(name,%22{allele_name}%22))\"\n",
    "    c_str+=s\n",
    "c_str = c_str[1:]\n",
    "\n",
    "c_str = prefix+c_str+\")\"\n",
    "c_str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['C*08:06', 'C*17:01', 'C*04:06']"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# checking which alleles were included from automatic searching and filtering\n",
    "# manually search and get remaining alleles\n",
    "c_out = pd.read_csv(\"HLA_C_rawOut.txt\", sep=\"\\t\")\n",
    "c_out \n",
    "\n",
    "c_not_included = []\n",
    "for c_allele in freq_alleles[\"C\"]:\n",
    "    # print(b_allele)\n",
    "    if c_out['name'].str.contains(c_allele).any() == False:\n",
    "        # print(b_allele)\n",
    "        c_not_included.append(c_allele)\n",
    "\n",
    "c_not_included\n",
    "\n",
    "# name_to_accession = dict(zip(b_out['name'], b_out['accession']))\n",
    "# name_to_accession"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### HLA-DQB1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'https://www.ebi.ac.uk/ipd/imgt/hla/alleles/?query=or(and(eq(cwd.version_2,%22C%22),startsWith(name,%22DQB1*03:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DQB1*02:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DQB1*05:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DQB1*03:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DQB1*05:03%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DQB1*06:04%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DQB1*04:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DQB1*06:09%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DQB1*06:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DQB1*03:03%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DQB1*04:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DQB1*06:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DQB1*05:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DQB1*06:03%22)))'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## only using CWD version 2 to filter seems to work well \n",
    "prefix = \"https://www.ebi.ac.uk/ipd/imgt/hla/alleles/?query=or(\"\n",
    "dqb1_str = str()\n",
    "for allele_name in freq_alleles[\"DQB1\"]:\n",
    "    #print(allele_name)\n",
    "    s = f\",and(eq(cwd.version_2,%22C%22),startsWith(name,%22{allele_name}%22))\"\n",
    "    dqb1_str+=s\n",
    "dqb1_str\n",
    "dqb1_str = dqb1_str[1:]\n",
    "\n",
    "dqb1_str = prefix+dqb1_str+\")\"\n",
    "dqb1_str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'DQB1*03:02:01:01': 'HLA00627',\n",
       " 'DQB1*02:01:01:01': 'HLA00622',\n",
       " 'DQB1*05:02:01:01': 'HLA00639',\n",
       " 'DQB1*03:01:01:01': 'HLA00625',\n",
       " 'DQB1*05:03:01:01': 'HLA00640',\n",
       " 'DQB1*06:04:01:01': 'HLA00648',\n",
       " 'DQB1*04:02:01:01': 'HLA00637',\n",
       " 'DQB1*06:09:01:01': 'HLA00654',\n",
       " 'DQB1*06:02:01:01': 'HLA00646',\n",
       " 'DQB1*03:03:02:01': 'HLA00629',\n",
       " 'DQB1*04:01:01:01': 'HLA00636',\n",
       " 'DQB1*06:01:01:01': 'HLA00643',\n",
       " 'DQB1*05:01:01:01': 'HLA00638',\n",
       " 'DQB1*06:03:01:01': 'HLA00647'}"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_to_a = pd.read_csv(\"HLA_name_to_dbAccession/HLA_DQB1_name_to_accession.txt\", sep=\"\\t\")\n",
    "n_to_a \n",
    "name_to_accession = dict(zip(n_to_a['name'], n_to_a['accession']))\n",
    "name_to_accession"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DQB1*03:02\n",
      "DQB1*02:01\n",
      "DQB1*05:02\n",
      "DQB1*03:01\n",
      "DQB1*05:03\n",
      "DQB1*06:04\n",
      "DQB1*04:02\n",
      "DQB1*06:09\n",
      "DQB1*06:02\n",
      "DQB1*03:03\n",
      "DQB1*04:01\n",
      "DQB1*06:01\n",
      "DQB1*05:01\n",
      "DQB1*06:03\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dqb1_out = pd.read_csv(\"HLA_name_to_dbAccession/HLA_DQB1_name_to_accession.txt\", sep=\"\\t\")\n",
    "\n",
    "# print(dqb1_out['name'].str.contains('DQB1*03:02', regex=False))\n",
    "not_included = []\n",
    "for dqb1_allele in freq_alleles[\"DQB1\"]:\n",
    "    print(dqb1_allele)\n",
    "    if dqb1_out['name'].str.contains(dqb1_allele, regex=False).any() == False:\n",
    "        print(dqb1_allele)\n",
    "        not_included.append(dqb1_allele)\n",
    "\n",
    "not_included"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### HLA-DRB1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'https://www.ebi.ac.uk/ipd/imgt/hla/alleles/?query=or(and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*14:05%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*14:04%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*15:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*04:05%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*08:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*04:07%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*15:06%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*13:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*09:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*08:04%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*11:04%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*14:08%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*04:06%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*15:03%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*13:03%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*04:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*12:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*04:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*01:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*11:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*14:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*15:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*03:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*14:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*03:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*01:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*12:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*14:03%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*13:04%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*08:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*04:04%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*14:06%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*13:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*13:12%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*11:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*01:03%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*16:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*04:11%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*04:10%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*07:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*10:01%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*04:03%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*16:02%22)),and(eq(cwd.version_2,%22C%22),startsWith(name,%22DRB1*08:03%22)))'"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## only using CWD version 2 to filter seems to work well \n",
    "prefix = \"https://www.ebi.ac.uk/ipd/imgt/hla/alleles/?query=or(\"\n",
    "drb1_str = str()\n",
    "for allele_name in freq_alleles[\"DRB1\"]:\n",
    "    #print(allele_name)\n",
    "    s = f\",and(eq(cwd.version_2,%22C%22),startsWith(name,%22{allele_name}%22))\"\n",
    "    drb1_str+=s\n",
    "drb1_str\n",
    "drb1_str = drb1_str[1:]\n",
    "\n",
    "drb1_str = prefix+drb1_str+\")\"\n",
    "drb1_str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "49"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_to_a = pd.read_csv(\"HLA_name_to_dbAccession/HLA_DRB1_name_to_accession.txt\", sep=\"\\t\")\n",
    "n_to_a \n",
    "name_to_accession = dict(zip(n_to_a['name'], n_to_a['accession']))\n",
    "name_to_accession\n",
    "len(name_to_accession)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DRB1*14:05\n",
      "DRB1*14:04\n",
      "DRB1*15:02\n",
      "DRB1*04:05\n",
      "DRB1*08:02\n",
      "DRB1*04:07\n",
      "DRB1*15:06\n",
      "DRB1*13:02\n",
      "DRB1*09:01\n",
      "DRB1*08:04\n",
      "DRB1*11:04\n",
      "DRB1*14:08\n",
      "DRB1*04:06\n",
      "DRB1*15:03\n",
      "DRB1*13:03\n",
      "DRB1*04:02\n",
      "DRB1*12:01\n",
      "DRB1*04:01\n",
      "DRB1*01:02\n",
      "DRB1*11:01\n",
      "DRB1*14:02\n",
      "DRB1*15:01\n",
      "DRB1*03:02\n",
      "DRB1*14:01\n",
      "DRB1*03:01\n",
      "DRB1*01:01\n",
      "DRB1*12:02\n",
      "DRB1*14:03\n",
      "DRB1*13:04\n",
      "DRB1*08:01\n",
      "DRB1*04:04\n",
      "DRB1*14:06\n",
      "DRB1*13:01\n",
      "DRB1*13:12\n",
      "DRB1*11:02\n",
      "DRB1*01:03\n",
      "DRB1*16:01\n",
      "DRB1*04:11\n",
      "DRB1*04:10\n",
      "DRB1*07:01\n",
      "DRB1*10:01\n",
      "DRB1*04:03\n",
      "DRB1*16:02\n",
      "DRB1*08:03\n",
      "not included []\n"
     ]
    }
   ],
   "source": [
    "drb1_out = pd.read_csv(\"HLA_name_to_dbAccession/HLA_DRB1_name_to_accession.txt\", sep=\"\\t\")\n",
    "\n",
    "# print(dqb1_out['name'].str.contains('DQB1*03:02', regex=False))\n",
    "not_included = []\n",
    "for drb1_allele in freq_alleles[\"DRB1\"]:\n",
    "    print(drb1_allele)\n",
    "    if drb1_out['name'].str.contains(drb1_allele, regex=False).any() == False:\n",
    "        print(drb1_allele)\n",
    "        not_included.append(drb1_allele)\n",
    "\n",
    "print(\"not included\", not_included)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "HLAoct2025",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
