{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## spliceR (python version) code"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import json\n",
    "from Bio.Seq import Seq\n",
    "import math\n",
    "import urllib.request"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "### reads information from json files from www.ebi.ac.uk HLA database\n",
    "### example webpage: https://www.ebi.ac.uk/ipd/imgt/hla/alleles/allele/?accession=HLA00050\n",
    "\n",
    "def parse_json(file_path=None, file_url=None):\n",
    "    \"\"\" takes in path to json file from www.ebi.ac.uk HLA database (default) OR url of json file \n",
    "    returns dict of:\n",
    "    - haplotype_name (str)\n",
    "    - prevalence (tuple of CWD v2.0 Status, CWD v3.0 Status (Total))\n",
    "    - genomic_features (list of dicts, one dict for each utr/exon/intron, keys are 'length', 'start', 'type', 'number' (intron/exons only))\n",
    "    - coding_features (similar to genomic features but only has exons - needed for pct cDNA calculation)\n",
    "    - sequences (dict of protein, coding, and genomic sequences (probably won't use for BE stuff))\n",
    "    - genomic_sequence (str)\n",
    "    \"\"\"\n",
    "    d = {}\n",
    "    if file_url is not None:\n",
    "        with urllib.request.urlopen(file_url) as url:\n",
    "            HLA_file_contents = json.load(url)\n",
    "    else: \n",
    "        with open(file_path) as HLA_file:\n",
    "            HLA_file_contents = json.load(HLA_file)\n",
    "        # keys we want: 'cwd' (for commonality), 'feature' (exon coords), 'name', 'sequence'\n",
    "    d['haplotype_name'] = HLA_file_contents['name']\n",
    "    # d['prevalence'] = (HLA_file_contents['cwd']['version_2'], HLA_file_contents['cwd']['version_3']['Total'])\n",
    "    d['genomic_features'] = HLA_file_contents['feature']['genomic']\n",
    "    d['coding_features'] = HLA_file_contents['feature']['coding']\n",
    "    d['sequences'] = HLA_file_contents['sequence']\n",
    "    d['genomic_sequence'] = HLA_file_contents['sequence']['genomic']\n",
    "\n",
    "    return d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### reads in motif and position weights from tsv files \n",
    "def parse_motif_position_weights(motif_weights_filepath, position_weights_filepath):\n",
    "    \"\"\" takes in filepath of tsv files storing motif weights and position weights, \n",
    "    returns 2 element tuple of motif_weights_dict, position_weights_dict\n",
    "        motif_weights_dict is dictionary mapping motif (7 NT str) to dictionary containing weight (float)\n",
    "        positions_weights_dict is nested dictionary mapping enzyme class (ABE or CBE) to dictionary mapping edited NT position to weight\n",
    "\n",
    "    example of using weight dictionaries: \n",
    "        pw = position_weights_dict[\"CBE\"][target_pos]['position_weight']\n",
    "        mw = motif_weights_dict[motif]['motif_weight']\n",
    "    \"\"\"\n",
    "    ## motif_weights_dict is dictionary mapping motif (7 NT str) to dictionary containing weight (float)\n",
    "    ## note that 4th NT must be A or C (for ABE or CBE base editing)\n",
    "    motif_weights_df = pd.read_csv(motif_weights_filepath, sep=\"\\t\", index_col=0)\n",
    "    motif_weights_dict = motif_weights_df.to_dict('index')\n",
    "\n",
    "    ## positions_weights_dict is nested dictionary mapping enzyme class (ABE or CBE) to dictionary mapping edited NT position to weight\n",
    "    position_weights_df = pd.read_csv(position_weights_filepath, sep=\"\\t\", index_col=0)\n",
    "    ABE_pw_df = position_weights_df[position_weights_df['enzyme_class'] == 'ABE']\n",
    "    ABE_pw_dict = ABE_pw_df.to_dict('index')\n",
    "    CBE_pw_df = position_weights_df[position_weights_df['enzyme_class'] == 'CBE']\n",
    "    CBE_pw_dict = CBE_pw_df.to_dict('index')\n",
    "    position_weights_dict = {\"ABE\": ABE_pw_dict, \"CBE\": CBE_pw_dict}\n",
    "\n",
    "    return motif_weights_dict, position_weights_dict\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "### functions for manipulating DNA seq strings and other basic helper functions\n",
    "\n",
    "def get_rev_complement(seq):\n",
    "    comp = str()\n",
    "    for char in seq:\n",
    "        if char == \"A\":\n",
    "            comp += \"T\"\n",
    "        elif char == \"T\":\n",
    "            comp += \"A\"\n",
    "        elif char == \"G\":\n",
    "            comp += \"C\"\n",
    "        elif char == \"C\":\n",
    "            comp += \"G\"\n",
    "    comp = comp[::-1]\n",
    "    return comp \n",
    "\n",
    "def matchDNAPattern(pattern, subject):\n",
    "    \"\"\" takes in a pattern (shorter DNA string) and subject (longer DNA string)\n",
    "    return list of 2 element lists:\n",
    "        each list giving start coord of matching region (in longer string) and subsequence of longer string that matches\n",
    "    note: \"N\" matches any of the 4 nucleotides (\"A\", \"T\", \"G\", \"C\")\n",
    "    \"\"\"\n",
    "    output = []\n",
    "    len_substring = len(pattern)\n",
    "    len_string = len(subject)\n",
    "    if len_substring>len_string:\n",
    "        raise Exception(\"pattern cannot be longer than subject\")\n",
    "    last_start = len_string-len_substring+1\n",
    "    for i in range(0,last_start):\n",
    "        # print(i)\n",
    "        si = subject[i:i+len_substring]\n",
    "        m = True\n",
    "        for idx, char in enumerate(si):\n",
    "            if (char == pattern[idx]) or (pattern[idx]==\"N\"):\n",
    "                continue\n",
    "            else:\n",
    "                m = False \n",
    "                break\n",
    "        if m:\n",
    "            output.append([i,si])\n",
    "    return(output)\n",
    "\n",
    "# test\n",
    "# print(matchDNAPattern(\"NNGCA\", \"ACTGCAGGGCA\"))\n",
    "# print(matchDNAPattern(\"NNGCAT\", \"ACTGCAGGGCA\"))\n",
    "# print(matchDNAPattern(\"NNGCAAAAAAAAAAAAAAT\", \"ACTGCAGGGCA\"))\n",
    "\n",
    "def probability(l):\n",
    "    return(math.exp(l)/(1+math.exp(l)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Main function generating dataframe with scored guides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "def spliceR_main(\n",
    "    genomic_features,\n",
    "    coding_features,\n",
    "    genomic_sequence,\n",
    "    coding_sequence,\n",
    "    motif_weights_dict,\n",
    "    position_weights_dict,\n",
    "    output_as_df = True,\n",
    "    haplotype_name = None,\n",
    "    guide_length = 20, \n",
    "    pam = \"NGG\",\n",
    "    flank_length = 20,\n",
    "    CBE = True, \n",
    "    ABE = True,\n",
    "    SA = True, \n",
    "    SD = True,\n",
    "    logistic_adjust = 1, \n",
    "    max_exon_targeted = None # for certain HLAs, we want to stop considering later exons as potential targets since they encode cytoplasmic tail/DQB1 has exon 5 splice acceptor polymorphisms which cause problems w the code \n",
    "):\n",
    "    \"\"\" \n",
    "    inputs: \n",
    "        - genomic_features: a list of dictionaries, one dict for each utr/exon/intron, keys are 'length', 'start', 'type', 'number' (for intron/exons only)\n",
    "        partial example: [{'length': 300, 'start': 1, 'type': '5utr'},\n",
    "        {'length': 73,\n",
    "        'number': '1',\n",
    "        'partial': False,\n",
    "        'start': 301,\n",
    "        'type': 'exon'},\n",
    "        {'length': 130,\n",
    "        'number': '1',\n",
    "        'partial': False,\n",
    "        'start': 374,\n",
    "        'type': 'intron'}, ...\n",
    "        - coding features: similar list of dictionaries as 'genomic_features' but only includes exons \n",
    "        - genomic sequence: string of genomic DNA sequence \n",
    "        - coding sequence: string of cDNA sequence \n",
    "    \n",
    "    returns: dataframe (or list of lists) with columns: \n",
    "    ['haplotype name', 'exon number', 'splice site type', 'protospacer', 'guide seq with PAM', 'PAM', 'enzyme class', 'CBE/ABE position', 'guide score', 'cDNA disruption pct']\n",
    "    \"\"\"\n",
    "    ### inner helper functions for getting CBE/ABE guides for SA and SD sites \n",
    "\n",
    "    def get_CBE_SD_guides(haplotype_name, exon_number, pam, exon_end_idx, flank_length, genomic_sequence, position_weights_dict, motif_weights_dict, guide_length, pct_of_cds):\n",
    "        pam_length = len(pam)\n",
    "        sd_start = exon_end_idx-flank_length-pam_length\n",
    "        sd_end = exon_end_idx+flank_length+pam_length\n",
    "        SD_seq = get_rev_complement(genomic_sequence[sd_start:sd_end])\n",
    "        guide_template_w_pam = \"N\"*guide_length+pam\n",
    "        unscored_CBE_SD_guides = matchDNAPattern(guide_template_w_pam, SD_seq) # 0 indexed \n",
    "        motif = get_rev_complement(genomic_sequence[exon_end_idx-3:exon_end_idx+4])\n",
    "        output = []\n",
    "        for u_cbe_sd_guide in unscored_CBE_SD_guides:\n",
    "            strt = u_cbe_sd_guide[0]\n",
    "            target_pos = flank_length+len(pam)-strt #position of target base within the guide (1 to 20 + pam)\n",
    "            if 1<=target_pos<=20:\n",
    "                pw = position_weights_dict[\"CBE\"][target_pos]['position_weight']\n",
    "                mw = motif_weights_dict[motif]['motif_weight']\n",
    "                cbe_score = probability(mw+pw-logistic_adjust)\n",
    "                u_cbe_sd_guide_wo_pam = u_cbe_sd_guide[1][:-3]\n",
    "                guide_info = [haplotype_name, exon_number, \"donor\",u_cbe_sd_guide_wo_pam, u_cbe_sd_guide[1], pam, \"CBE\", target_pos, cbe_score, pct_of_cds]\n",
    "                output.append(guide_info)\n",
    "        return output\n",
    "\n",
    "    def get_ABE_SD_guides(haplotype_name, exon_number, pam, exon_end_idx, flank_length, genomic_sequence, position_weights_dict, motif_weights_dict, guide_length, pct_of_cds):\n",
    "        pam_length = len(pam)\n",
    "        sd_start = exon_end_idx-flank_length-pam_length\n",
    "        sd_end = exon_end_idx+flank_length+pam_length\n",
    "        SD_seq = get_rev_complement(genomic_sequence[sd_start:sd_end])\n",
    "        guide_template_w_pam = \"N\"*guide_length+pam\n",
    "        unscored_ABE_SD_guides = matchDNAPattern(guide_template_w_pam, SD_seq) # 0 indexed \n",
    "        motif = get_rev_complement(genomic_sequence[exon_end_idx-2:exon_end_idx+5])\n",
    "        output = []\n",
    "        for u_abe_sd_guide in unscored_ABE_SD_guides:\n",
    "            strt = u_abe_sd_guide[0]\n",
    "            target_pos = flank_length+len(pam)-strt-1 #position of target base within the guide (1 to 20 + pam)\n",
    "            if 1<=target_pos<=20:\n",
    "                pw = position_weights_dict[\"ABE\"][target_pos]['position_weight']\n",
    "                mw = motif_weights_dict[motif]['motif_weight']\n",
    "                abe_score = probability(mw+pw-logistic_adjust)\n",
    "                u_abe_sd_guide_wo_pam = u_abe_sd_guide[1][:-3]\n",
    "                guide_info = [haplotype_name, exon_number, \"donor\", u_abe_sd_guide_wo_pam, u_abe_sd_guide[1], pam, \"ABE\", target_pos, abe_score, pct_of_cds]\n",
    "                output.append(guide_info)\n",
    "        return output\n",
    "\n",
    "    def get_CBE_SA_guides(haplotype_name, exon_number, pam, exon_start_idx, flank_length, genomic_sequence, position_weights_dict, motif_weights_dict, guide_length, pct_of_cds):\n",
    "        pam_length = len(pam)\n",
    "        sa_start = exon_start_idx-flank_length-pam_length\n",
    "        sa_end = exon_start_idx+flank_length+pam_length\n",
    "        SA_seq = get_rev_complement(genomic_sequence[sa_start:sa_end])\n",
    "        guide_template_w_pam = \"N\"*guide_length+pam\n",
    "        unscored_CBE_SA_guides = matchDNAPattern(guide_template_w_pam, SA_seq)\n",
    "        motif = get_rev_complement(genomic_sequence[exon_start_idx-4:exon_start_idx+3]) #checked \n",
    "        # scoring all potential guides\n",
    "        output = []\n",
    "        for u_cbe_sa_guide in unscored_CBE_SA_guides: \n",
    "            strt = u_cbe_sa_guide[0]\n",
    "            target_pos = flank_length+len(pam)-strt+1 #position of target base within the guide (1 to 20 + pam) checked\n",
    "            if 1<=target_pos<=20:\n",
    "                pw = position_weights_dict[\"CBE\"][target_pos]['position_weight']\n",
    "                mw = motif_weights_dict[motif]['motif_weight']\n",
    "                cbe_score = probability(mw+pw-logistic_adjust)\n",
    "                u_cbe_sa_guide_wo_pam = u_cbe_sa_guide[1][:-3]\n",
    "                guide_info = [haplotype_name, exon_number, \"acceptor\", u_cbe_sa_guide_wo_pam, u_cbe_sa_guide[1], pam, \"CBE\", target_pos, cbe_score, pct_of_cds]\n",
    "                output.append(guide_info)\n",
    "        return output\n",
    "\n",
    "    def get_ABE_SA_guides(haplotype_name, exon_number, pam, exon_start_idx, flank_length, genomic_sequence, position_weights_dict, motif_weights_dict, guide_length, pct_of_cds):\n",
    "        # unlike the other 3 cases, SA seq to be edited is NOT reverse complement \n",
    "        pam_length = len(pam)\n",
    "        sa_start = exon_start_idx-flank_length-pam_length\n",
    "        sa_end = exon_start_idx+flank_length+pam_length\n",
    "        guide_template_w_pam = \"N\"*guide_length+pam\n",
    "        SA_seq = genomic_sequence[sa_start:sa_end]\n",
    "        unscored_ABE_SA_guides = matchDNAPattern(guide_template_w_pam, SA_seq)\n",
    "        motif = genomic_sequence[exon_start_idx-5:exon_start_idx+2] #checked\n",
    "        # scoring all potential guides\n",
    "        output=[]\n",
    "        for u_abe_sa_guide in unscored_ABE_SA_guides: \n",
    "            strt = u_abe_sa_guide[0]\n",
    "            target_pos = flank_length+len(pam)-strt-1 #position of target base within the guide (1 to 20 + pam) checked\n",
    "            if 1<=target_pos<=20:\n",
    "                pw = position_weights_dict[\"ABE\"][target_pos]['position_weight']\n",
    "                mw = motif_weights_dict[motif]['motif_weight']\n",
    "                abe_score = probability(mw+pw-logistic_adjust)\n",
    "                u_abe_sa_guide_wo_pam = u_abe_sa_guide[1][:-3]\n",
    "                guide_info = [haplotype_name, exon_number, \"acceptor\", u_abe_sa_guide_wo_pam, u_abe_sa_guide[1], pam, \"ABE\", target_pos, abe_score, pct_of_cds]\n",
    "                output.append(guide_info)\n",
    "        return output\n",
    "    \n",
    "    #### MAIN FUNCTION \n",
    "    cols = ['haplotype name', 'exon number', 'splice site type', 'protospacer', 'guide seq with PAM', 'PAM', 'enzyme class', 'CBE/ABE position', 'guide score', 'cDNA disruption pct']\n",
    "\n",
    "    exons_list = []\n",
    "    num_exons = 0 \n",
    "    for feature in genomic_features:\n",
    "        if feature['type'] =='exon':\n",
    "            if max_exon_targeted is not None:\n",
    "                if int(feature['number']) <= max_exon_targeted:\n",
    "                    exons_list.append(feature)\n",
    "            else: exons_list.append(feature)\n",
    "            if int(feature['number'])>num_exons:\n",
    "                num_exons=int(feature['number'])\n",
    "\n",
    "    print(\"num exons\", num_exons)\n",
    "    ## list to collect all outputs \n",
    "    output_list_of_lists = [] # each list corresponds to a possible guide \n",
    "    ## loop over all exons \n",
    "    for exon in exons_list:\n",
    "        print(\"exon\", exon['number'])\n",
    "\n",
    "        ## calculating what percentage of the way through coding sequence the end of this exon is (want earlier disruption)\n",
    "        end_of_exon_in_cds = 0 \n",
    "        for cd_feat in coding_features:\n",
    "            if cd_feat['number'] == exon['number']:\n",
    "                end_of_exon_in_cds=cd_feat['start']+cd_feat['length']\n",
    "                break\n",
    "        pct_of_cds = end_of_exon_in_cds/len(coding_sequence)\n",
    "        \n",
    "        ## getting start and end coordinates of this exon (relative to GENOMIC sequence)\n",
    "        exon_start_idx = exon['start']-1 #-1 to account for 0 indexing \n",
    "        exon_end_idx = exon_start_idx+exon['length']\n",
    "\n",
    "        # check if it's first exon, if so, only get splice donor (after exon)\n",
    "        if int(exon['number']) == 1:\n",
    "            if SD:\n",
    "                if CBE:\n",
    "                    output_list_of_lists.extend(get_CBE_SD_guides(haplotype_name, exon['number'], pam, exon_end_idx, flank_length, genomic_sequence, position_weights_dict, motif_weights_dict, guide_length, pct_of_cds))\n",
    "                \n",
    "                if ABE:\n",
    "                    output_list_of_lists.extend(get_ABE_SD_guides(haplotype_name, exon['number'], pam, exon_end_idx, flank_length, genomic_sequence, position_weights_dict, motif_weights_dict, guide_length, pct_of_cds))\n",
    "\n",
    "        # check if it's last exon, if so, only get splice acceptor (before exon)\n",
    "        elif int(exon['number']) == num_exons:\n",
    "            # print(\"last exon\", int(exon['number']))\n",
    "            if SA:\n",
    "                if CBE:\n",
    "                    output_list_of_lists.extend(get_CBE_SA_guides(haplotype_name, exon['number'], pam, exon_start_idx, flank_length, genomic_sequence, position_weights_dict, motif_weights_dict, guide_length, pct_of_cds))\n",
    "\n",
    "                if ABE: \n",
    "                    output_list_of_lists.extend(get_ABE_SA_guides(haplotype_name, exon['number'], pam, exon_start_idx, flank_length, genomic_sequence, position_weights_dict, motif_weights_dict, guide_length, pct_of_cds))\n",
    "        \n",
    "        else:\n",
    "            # splice acceptor sites \n",
    "            if SA:\n",
    "                if CBE:\n",
    "                    output_list_of_lists.extend(get_CBE_SA_guides(haplotype_name, exon['number'], pam, exon_start_idx, flank_length, genomic_sequence, position_weights_dict, motif_weights_dict, guide_length, pct_of_cds))\n",
    "                    print(\"SA, CBE\")\n",
    "                if ABE: \n",
    "                    output_list_of_lists.extend(get_ABE_SA_guides(haplotype_name, exon['number'], pam, exon_start_idx, flank_length, genomic_sequence, position_weights_dict, motif_weights_dict, guide_length, pct_of_cds))\n",
    "                    print(\"SA, ABE\")\n",
    "            # splice donor sites \n",
    "            if SD:\n",
    "                if CBE:\n",
    "                    output_list_of_lists.extend(get_CBE_SD_guides(haplotype_name, exon['number'], pam, exon_end_idx, flank_length, genomic_sequence, position_weights_dict, motif_weights_dict, guide_length, pct_of_cds))\n",
    "                    print(\"SD, CBE\")\n",
    "                if ABE:\n",
    "                    output_list_of_lists.extend(get_ABE_SD_guides(haplotype_name, exon['number'], pam, exon_end_idx, flank_length, genomic_sequence, position_weights_dict, motif_weights_dict, guide_length, pct_of_cds))\n",
    "                    print(\"SD, ABE\")\n",
    "    \n",
    "    if output_as_df:\n",
    "        return pd.DataFrame(output_list_of_lists, columns=cols)\n",
    "    \n",
    "    else: return output_list_of_lists\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## For HLA project"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### HLA-A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "40"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_to_a = pd.read_csv(\"HLA_A_name_to_accession.txt\", sep=\"\\t\")\n",
    "n_to_a \n",
    "HLA_A_name_to_accession = dict(zip(n_to_a['name'], n_to_a['accession']))\n",
    "HLA_A_name_to_accession\n",
    "len(HLA_A_name_to_accession)\n",
    "# name_to_accession #dict mapping HLA haplotype name to accession "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A*30:04:01:01\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/sw/t72qsf8n6dj3nspvp9_rlmxr0000gp/T/ipykernel_57621/3182564382.py:40: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  HLA_A_output_df = pd.concat([HLA_A_output_df, hap_out], axis=0, ignore_index=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A*24:03:01:01\n",
      "A*24:02:01:01\n",
      "A*11:01:01:01\n",
      "A*34:02:01:01\n",
      "A*32:01:01:01\n",
      "A*01:01:01:01\n",
      "A*03:02:01:01\n",
      "A*68:01:01:01\n",
      "A*68:01:02:01\n",
      "A*23:01:01:01\n",
      "A*11:02:01:01\n",
      "A*74:01:01:01\n",
      "A*02:05:01:01\n",
      "A*26:01:01:01\n",
      "A*30:01:01:01\n",
      "A*34:01:01:01\n",
      "A*26:03:01:01\n",
      "A*66:01:01:01\n",
      "A*03:01:01:01\n",
      "A*02:11:01:01\n",
      "A*02:02:01:01\n",
      "A*02:03:01:01\n",
      "A*33:01:01:01\n",
      "A*30:02:01:01\n",
      "A*24:07:01:01\n",
      "A*29:02:01:01\n",
      "A*25:01:01:01\n",
      "A*68:02:01:01\n",
      "A*33:03:01:01\n",
      "A*36:01:01:01\n",
      "A*29:01:01:01\n",
      "A*02:01:01:01\n",
      "A*02:01:04\n",
      "A*31:01:02:01\n",
      "A*68:03:01:01\n",
      "A*02:06:01:01\n",
      "A*02:07:01:01\n",
      "A*26:02:01\n",
      "A*02:17:02:01\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>haplotype name</th>\n",
       "      <th>exon number</th>\n",
       "      <th>splice site type</th>\n",
       "      <th>protospacer</th>\n",
       "      <th>guide seq with PAM</th>\n",
       "      <th>PAM</th>\n",
       "      <th>enzyme class</th>\n",
       "      <th>CBE/ABE position</th>\n",
       "      <th>guide score</th>\n",
       "      <th>cDNA distuption pct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A*30:04:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>GACCCCGCACTCACCCGCCC</td>\n",
       "      <td>GACCCCGCACTCACCCGCCCAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>14</td>\n",
       "      <td>0.113088</td>\n",
       "      <td>0.067395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A*30:04:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>CGCACTCACCCGCCCAGGTC</td>\n",
       "      <td>CGCACTCACCCGCCCAGGTCTGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>9</td>\n",
       "      <td>0.334910</td>\n",
       "      <td>0.067395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A*30:04:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>GCACTCACCCGCCCAGGTCT</td>\n",
       "      <td>GCACTCACCCGCCCAGGTCTGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>8</td>\n",
       "      <td>0.466535</td>\n",
       "      <td>0.067395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A*30:04:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>CACCCGCCCAGGTCTGGGTC</td>\n",
       "      <td>CACCCGCCCAGGTCTGGGTCAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>3</td>\n",
       "      <td>0.336702</td>\n",
       "      <td>0.067395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A*30:04:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>ACCCGCCCAGGTCTGGGTCA</td>\n",
       "      <td>ACCCGCCCAGGTCTGGGTCAGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>2</td>\n",
       "      <td>0.257234</td>\n",
       "      <td>0.067395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2011</th>\n",
       "      <td>A*02:17:02:01</td>\n",
       "      <td>7</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>CCCTGGGCACTGTCACTGCC</td>\n",
       "      <td>CCCTGGGCACTGTCACTGCCTGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>20</td>\n",
       "      <td>0.044522</td>\n",
       "      <td>0.996357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2012</th>\n",
       "      <td>A*02:17:02:01</td>\n",
       "      <td>7</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>CCTGGGCACTGTCACTGCCT</td>\n",
       "      <td>CCTGGGCACTGTCACTGCCTGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>19</td>\n",
       "      <td>0.059258</td>\n",
       "      <td>0.996357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2013</th>\n",
       "      <td>A*02:17:02:01</td>\n",
       "      <td>7</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>CTGGGCACTGTCACTGCCTG</td>\n",
       "      <td>CTGGGCACTGTCACTGCCTGGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>18</td>\n",
       "      <td>0.085840</td>\n",
       "      <td>0.996357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014</th>\n",
       "      <td>A*02:17:02:01</td>\n",
       "      <td>7</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>CCCAGGCAGTGACAGTGCCC</td>\n",
       "      <td>CCCAGGCAGTGACAGTGCCCAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>4</td>\n",
       "      <td>0.238371</td>\n",
       "      <td>0.996357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2015</th>\n",
       "      <td>A*02:17:02:01</td>\n",
       "      <td>7</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>CCAGGCAGTGACAGTGCCCA</td>\n",
       "      <td>CCAGGCAGTGACAGTGCCCAGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>3</td>\n",
       "      <td>0.112914</td>\n",
       "      <td>0.996357</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2016 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     haplotype name exon number splice site type           protospacer  \\\n",
       "0     A*30:04:01:01           1            donor  GACCCCGCACTCACCCGCCC   \n",
       "1     A*30:04:01:01           1            donor  CGCACTCACCCGCCCAGGTC   \n",
       "2     A*30:04:01:01           1            donor  GCACTCACCCGCCCAGGTCT   \n",
       "3     A*30:04:01:01           1            donor  CACCCGCCCAGGTCTGGGTC   \n",
       "4     A*30:04:01:01           1            donor  ACCCGCCCAGGTCTGGGTCA   \n",
       "...             ...         ...              ...                   ...   \n",
       "2011  A*02:17:02:01           7         acceptor  CCCTGGGCACTGTCACTGCC   \n",
       "2012  A*02:17:02:01           7         acceptor  CCTGGGCACTGTCACTGCCT   \n",
       "2013  A*02:17:02:01           7         acceptor  CTGGGCACTGTCACTGCCTG   \n",
       "2014  A*02:17:02:01           7         acceptor  CCCAGGCAGTGACAGTGCCC   \n",
       "2015  A*02:17:02:01           7         acceptor  CCAGGCAGTGACAGTGCCCA   \n",
       "\n",
       "           guide seq with PAM  PAM enzyme class CBE/ABE position  guide score  \\\n",
       "0     GACCCCGCACTCACCCGCCCAGG  NGG          CBE               14     0.113088   \n",
       "1     CGCACTCACCCGCCCAGGTCTGG  NGG          CBE                9     0.334910   \n",
       "2     GCACTCACCCGCCCAGGTCTGGG  NGG          CBE                8     0.466535   \n",
       "3     CACCCGCCCAGGTCTGGGTCAGG  NGG          CBE                3     0.336702   \n",
       "4     ACCCGCCCAGGTCTGGGTCAGGG  NGG          CBE                2     0.257234   \n",
       "...                       ...  ...          ...              ...          ...   \n",
       "2011  CCCTGGGCACTGTCACTGCCTGG  NGG          CBE               20     0.044522   \n",
       "2012  CCTGGGCACTGTCACTGCCTGGG  NGG          CBE               19     0.059258   \n",
       "2013  CTGGGCACTGTCACTGCCTGGGG  NGG          CBE               18     0.085840   \n",
       "2014  CCCAGGCAGTGACAGTGCCCAGG  NGG          ABE                4     0.238371   \n",
       "2015  CCAGGCAGTGACAGTGCCCAGGG  NGG          ABE                3     0.112914   \n",
       "\n",
       "      cDNA distuption pct  \n",
       "0                0.067395  \n",
       "1                0.067395  \n",
       "2                0.067395  \n",
       "3                0.067395  \n",
       "4                0.067395  \n",
       "...                   ...  \n",
       "2011             0.996357  \n",
       "2012             0.996357  \n",
       "2013             0.996357  \n",
       "2014             0.996357  \n",
       "2015             0.996357  \n",
       "\n",
       "[2016 rows x 10 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#### running spliceR on all HLA A common haplotypes \n",
    "\n",
    "# initializing empty dataframe to store results \n",
    "cols = ['haplotype name', 'exon number', 'splice site type', 'protospacer', 'guide seq with PAM', 'PAM', 'enzyme class', 'CBE/ABE position', 'guide score', 'cDNA disruption pct']\n",
    "HLA_A_output_df = pd.DataFrame(columns=cols)\n",
    "\n",
    "motif_weights_filepath = \"./motif_weights.tsv\"\n",
    "position_weights_filepath = \"./position_weights.tsv\"\n",
    "motif_weights_dict, position_weights_dict = parse_motif_position_weights(motif_weights_filepath, position_weights_filepath)\n",
    "\n",
    "for haplotype in HLA_A_name_to_accession:\n",
    "    print(haplotype)\n",
    "    accession = HLA_A_name_to_accession[haplotype]\n",
    "    hap_url = f\"https://www.ebi.ac.uk/cgi-bin/ipd/api/allele/{accession}\"\n",
    "    # print(hp_url)\n",
    "    hap_info = parse_json(file_url=hap_url)\n",
    "    genomic_features = hap_info['genomic_features']\n",
    "    coding_features = hap_info['coding_features']\n",
    "    genomic_sequence = hap_info['genomic_sequence']\n",
    "    coding_sequence = hap_info['sequences']['coding']\n",
    "\n",
    "    hap_out = spliceR_main(\n",
    "        genomic_features,\n",
    "        coding_features,\n",
    "        genomic_sequence,\n",
    "        coding_sequence,\n",
    "        motif_weights_dict,\n",
    "        position_weights_dict,\n",
    "        output_as_df = True,\n",
    "        haplotype_name = haplotype,\n",
    "        guide_length = 20, \n",
    "        pam = \"NGG\",\n",
    "        flank_length = 20,\n",
    "        CBE = True, \n",
    "        ABE = True,\n",
    "        SA = True, \n",
    "        SD = True,\n",
    "        logistic_adjust = 1\n",
    "    )\n",
    "    HLA_A_output_df = pd.concat([HLA_A_output_df, hap_out], axis=0, ignore_index=True)\n",
    "\n",
    "\n",
    "HLA_A_output_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "HLA_A_output_df.to_csv('HLA_A_all_haps_splice_site_BE.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### HLA-B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "76"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hla_b_n_to_a = pd.read_csv(\"HLA_B_name_to_accession.txt\", sep=\"\\t\")\n",
    "hla_b_n_to_a \n",
    "HLA_B_name_to_accession = dict(zip(hla_b_n_to_a['name'], hla_b_n_to_a['accession']))\n",
    "HLA_B_name_to_accession\n",
    "len(HLA_B_name_to_accession)\n",
    "\n",
    "## there are 76 different haplotypes to target b/c ig 2 of the original list of 74 have 2 versions that are both marked as common?? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "B*54:01:01:01\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/sw/t72qsf8n6dj3nspvp9_rlmxr0000gp/T/ipykernel_81124/2423316822.py:40: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  HLA_B_output_df = pd.concat([HLA_B_output_df, hap_out], axis=0, ignore_index=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "B*55:01:01:01\n",
      "B*59:01:01:01\n",
      "B*44:02:01:01\n",
      "B*81:01:01:01\n",
      "B*13:02:01:01\n",
      "B*07:02:01:01\n",
      "B*07:05:01:01\n",
      "B*50:01:01:01\n",
      "B*35:12:01:01\n",
      "B*51:06:01:01\n",
      "B*13:01:01:01\n",
      "B*41:01:01:01\n",
      "B*15:11:01:01\n",
      "B*39:06:02:01\n",
      "B*57:01:01:01\n",
      "B*44:03:01:01\n",
      "B*44:03:02:01\n",
      "B*51:02:01:01\n",
      "B*15:35:01:01\n",
      "B*56:01:01:01\n",
      "B*15:16:01:01\n",
      "B*57:03:01:01\n",
      "B*27:04:01\n",
      "B*35:01:01:01\n",
      "B*27:05:02:01\n",
      "B*58:02:01:01\n",
      "B*58:01:01:01\n",
      "B*42:01:01:01\n",
      "B*15:01:01:01\n",
      "B*15:15:01:01\n",
      "B*39:10:01\n",
      "B*15:03:01:01\n",
      "B*14:02:01:01\n",
      "B*49:01:01:01\n",
      "B*35:02:01:01\n",
      "B*51:01:01:01\n",
      "B*46:01:01:01\n",
      "B*35:08:01:01\n",
      "B*15:21:01:01\n",
      "B*55:02:01:01\n",
      "B*40:02:01:01\n",
      "B*15:13:01\n",
      "B*35:17:01:01\n",
      "B*53:01:01:01\n",
      "B*15:10:01:01\n",
      "B*38:01:01:01\n",
      "B*15:12:01\n",
      "B*37:01:01:01\n",
      "B*35:43:01\n",
      "B*27:06:01:01\n",
      "B*35:03:01:01\n",
      "B*18:01:01:01\n",
      "B*48:01:01:01\n",
      "B*35:05:01:01\n",
      "B*15:17:01:01\n",
      "B*40:06:01:01\n",
      "B*52:01:01:01\n",
      "B*52:01:02:01\n",
      "B*15:18:01:01\n",
      "B*14:01:01:01\n",
      "B*38:02:01:01\n",
      "B*39:01:01:01\n",
      "B*15:25:01:01\n",
      "B*39:05:01:01\n",
      "B*08:01:01:01\n",
      "B*45:01:01:01\n",
      "B*67:01:01\n",
      "B*15:02:01:01\n",
      "B*40:01:02:01\n",
      "B*15:07:01:01\n",
      "B*40:10:01:01\n",
      "B*15:06\n",
      "B*56:02:01:01\n",
      "B*56:04:01:01\n",
      "B*15:27:01:01\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>haplotype name</th>\n",
       "      <th>exon number</th>\n",
       "      <th>splice site type</th>\n",
       "      <th>protospacer</th>\n",
       "      <th>guide seq with PAM</th>\n",
       "      <th>PAM</th>\n",
       "      <th>enzyme class</th>\n",
       "      <th>CBE/ABE position</th>\n",
       "      <th>guide score</th>\n",
       "      <th>cDNA distuption pct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>B*54:01:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>CCCTCCCGACCCGCACTCAC</td>\n",
       "      <td>CCCTCCCGACCCGCACTCACCGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>20</td>\n",
       "      <td>0.044239</td>\n",
       "      <td>0.067952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>B*54:01:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>CGACCCGCACTCACCGGCCC</td>\n",
       "      <td>CGACCCGCACTCACCGGCCCAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>14</td>\n",
       "      <td>0.143533</td>\n",
       "      <td>0.067952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>B*54:01:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>GCACTCACCGGCCCAGGTCT</td>\n",
       "      <td>GCACTCACCGGCCCAGGTCTCGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>8</td>\n",
       "      <td>0.534763</td>\n",
       "      <td>0.067952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>B*54:01:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>CACCGGCCCAGGTCTCGGTC</td>\n",
       "      <td>CACCGGCCCAGGTCTCGGTCAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>3</td>\n",
       "      <td>0.400186</td>\n",
       "      <td>0.067952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>B*54:01:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>ACCGGCCCAGGTCTCGGTCA</td>\n",
       "      <td>ACCGGCCCAGGTCTCGGTCAGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>2</td>\n",
       "      <td>0.312800</td>\n",
       "      <td>0.067952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4208</th>\n",
       "      <td>B*15:27:01:01</td>\n",
       "      <td>6</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>TTCCCACAGGTGGAAAAGGA</td>\n",
       "      <td>TTCCCACAGGTGGAAAAGGAGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>8</td>\n",
       "      <td>0.254955</td>\n",
       "      <td>0.960514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4209</th>\n",
       "      <td>B*15:27:01:01</td>\n",
       "      <td>7</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>CCCTGGGCACTGTCGCTGGC</td>\n",
       "      <td>CCCTGGGCACTGTCGCTGGCTGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>20</td>\n",
       "      <td>0.006267</td>\n",
       "      <td>1.000918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4210</th>\n",
       "      <td>B*15:27:01:01</td>\n",
       "      <td>7</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>GGCTGGAGTAGAACAAAAAC</td>\n",
       "      <td>GGCTGGAGTAGAACAAAAACAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>3</td>\n",
       "      <td>0.083324</td>\n",
       "      <td>1.000918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4211</th>\n",
       "      <td>B*15:27:01:01</td>\n",
       "      <td>7</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>TCCAGCCAGCGACAGTGCCC</td>\n",
       "      <td>TCCAGCCAGCGACAGTGCCCAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>4</td>\n",
       "      <td>0.145725</td>\n",
       "      <td>1.000918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4212</th>\n",
       "      <td>B*15:27:01:01</td>\n",
       "      <td>7</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>CCAGCCAGCGACAGTGCCCA</td>\n",
       "      <td>CCAGCCAGCGACAGTGCCCAGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>3</td>\n",
       "      <td>0.064875</td>\n",
       "      <td>1.000918</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4213 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     haplotype name exon number splice site type           protospacer  \\\n",
       "0     B*54:01:01:01           1            donor  CCCTCCCGACCCGCACTCAC   \n",
       "1     B*54:01:01:01           1            donor  CGACCCGCACTCACCGGCCC   \n",
       "2     B*54:01:01:01           1            donor  GCACTCACCGGCCCAGGTCT   \n",
       "3     B*54:01:01:01           1            donor  CACCGGCCCAGGTCTCGGTC   \n",
       "4     B*54:01:01:01           1            donor  ACCGGCCCAGGTCTCGGTCA   \n",
       "...             ...         ...              ...                   ...   \n",
       "4208  B*15:27:01:01           6         acceptor  TTCCCACAGGTGGAAAAGGA   \n",
       "4209  B*15:27:01:01           7         acceptor  CCCTGGGCACTGTCGCTGGC   \n",
       "4210  B*15:27:01:01           7         acceptor  GGCTGGAGTAGAACAAAAAC   \n",
       "4211  B*15:27:01:01           7         acceptor  TCCAGCCAGCGACAGTGCCC   \n",
       "4212  B*15:27:01:01           7         acceptor  CCAGCCAGCGACAGTGCCCA   \n",
       "\n",
       "           guide seq with PAM  PAM enzyme class CBE/ABE position  guide score  \\\n",
       "0     CCCTCCCGACCCGCACTCACCGG  NGG          CBE               20     0.044239   \n",
       "1     CGACCCGCACTCACCGGCCCAGG  NGG          CBE               14     0.143533   \n",
       "2     GCACTCACCGGCCCAGGTCTCGG  NGG          CBE                8     0.534763   \n",
       "3     CACCGGCCCAGGTCTCGGTCAGG  NGG          CBE                3     0.400186   \n",
       "4     ACCGGCCCAGGTCTCGGTCAGGG  NGG          CBE                2     0.312800   \n",
       "...                       ...  ...          ...              ...          ...   \n",
       "4208  TTCCCACAGGTGGAAAAGGAGGG  NGG          ABE                8     0.254955   \n",
       "4209  CCCTGGGCACTGTCGCTGGCTGG  NGG          CBE               20     0.006267   \n",
       "4210  GGCTGGAGTAGAACAAAAACAGG  NGG          CBE                3     0.083324   \n",
       "4211  TCCAGCCAGCGACAGTGCCCAGG  NGG          ABE                4     0.145725   \n",
       "4212  CCAGCCAGCGACAGTGCCCAGGG  NGG          ABE                3     0.064875   \n",
       "\n",
       "      cDNA distuption pct  \n",
       "0                0.067952  \n",
       "1                0.067952  \n",
       "2                0.067952  \n",
       "3                0.067952  \n",
       "4                0.067952  \n",
       "...                   ...  \n",
       "4208             0.960514  \n",
       "4209             1.000918  \n",
       "4210             1.000918  \n",
       "4211             1.000918  \n",
       "4212             1.000918  \n",
       "\n",
       "[4213 rows x 10 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#### running spliceR on all HLA B common haplotypes \n",
    "\n",
    "# initializing empty dataframe to store results \n",
    "cols = ['haplotype name', 'exon number', 'splice site type', 'protospacer', 'guide seq with PAM', 'PAM', 'enzyme class', 'CBE/ABE position', 'guide score', 'cDNA disruption pct']\n",
    "HLA_B_output_df = pd.DataFrame(columns=cols)\n",
    "\n",
    "motif_weights_filepath = \"./motif_weights.tsv\"\n",
    "position_weights_filepath = \"./position_weights.tsv\"\n",
    "motif_weights_dict, position_weights_dict = parse_motif_position_weights(motif_weights_filepath, position_weights_filepath)\n",
    "\n",
    "for haplotype in HLA_B_name_to_accession:\n",
    "    print(haplotype)\n",
    "    accession = HLA_B_name_to_accession[haplotype]\n",
    "    hap_url = f\"https://www.ebi.ac.uk/cgi-bin/ipd/api/allele/{accession}\"\n",
    "    # print(hp_url)\n",
    "    hap_info = parse_json(file_url=hap_url)\n",
    "    genomic_features = hap_info['genomic_features']\n",
    "    coding_features = hap_info['coding_features']\n",
    "    genomic_sequence = hap_info['genomic_sequence']\n",
    "    coding_sequence = hap_info['sequences']['coding']\n",
    "\n",
    "    hap_out = spliceR_main(\n",
    "        genomic_features,\n",
    "        coding_features,\n",
    "        genomic_sequence,\n",
    "        coding_sequence,\n",
    "        motif_weights_dict,\n",
    "        position_weights_dict,\n",
    "        output_as_df = True,\n",
    "        haplotype_name = haplotype,\n",
    "        guide_length = 20, \n",
    "        pam = \"NGG\",\n",
    "        flank_length = 20,\n",
    "        CBE = True, \n",
    "        ABE = True,\n",
    "        SA = True, \n",
    "        SD = True,\n",
    "        logistic_adjust = 1\n",
    "    )\n",
    "    HLA_B_output_df = pd.concat([HLA_B_output_df, hap_out], axis=0, ignore_index=True)\n",
    "\n",
    "\n",
    "HLA_B_output_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "HLA_B_output_df.to_csv('HLA_B_all_haps_splice_site_BE.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### HLA-C "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hla_c_n_to_a = pd.read_csv(\"HLA_C_name_to_accession.txt\", sep=\"\\t\")\n",
    "hla_c_n_to_a \n",
    "HLA_C_name_to_accession = dict(zip(hla_c_n_to_a['name'], hla_c_n_to_a['accession']))\n",
    "HLA_C_name_to_accession\n",
    "len(HLA_C_name_to_accession) # 35 target haplotypes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C*07:02:01:01\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/sw/t72qsf8n6dj3nspvp9_rlmxr0000gp/T/ipykernel_81124/1427478302.py:40: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  HLA_C_output_df = pd.concat([HLA_C_output_df, hap_out], axis=0, ignore_index=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C*14:02:01:01\n",
      "C*14:02:03\n",
      "C*07:04:01:01\n",
      "C*08:04:01:01\n",
      "C*08:01:01:01\n",
      "C*15:02:01:01\n",
      "C*03:04:01:01\n",
      "C*03:04:02:01\n",
      "C*12:04:02:01\n",
      "C*03:02:01\n",
      "C*08:03:01:01\n",
      "C*04:01:01:01\n",
      "C*07:01:01:01\n",
      "C*16:01:01:01\n",
      "C*06:02:01:01\n",
      "C*05:01:01:01\n",
      "C*03:03:01:01\n",
      "C*03:03:04\n",
      "C*12:02:02:01\n",
      "C*03:06:01:01\n",
      "C*16:02:01:01\n",
      "C*01:02:01:01\n",
      "C*03:05:01:01\n",
      "C*14:03:01:01\n",
      "C*02:02:02:01\n",
      "C*16:04:01:01\n",
      "C*18:01:01:01\n",
      "C*04:03:01:01\n",
      "C*15:05:01:01\n",
      "C*12:03:01:01\n",
      "C*08:02:01:01\n",
      "C*08:06\n",
      "C*17:01:01:02\n",
      "C*04:06:01\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>haplotype name</th>\n",
       "      <th>exon number</th>\n",
       "      <th>splice site type</th>\n",
       "      <th>protospacer</th>\n",
       "      <th>guide seq with PAM</th>\n",
       "      <th>PAM</th>\n",
       "      <th>enzyme class</th>\n",
       "      <th>CBE/ABE position</th>\n",
       "      <th>guide score</th>\n",
       "      <th>cDNA distuption pct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C*07:02:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>CCTCCCAACCCCGCACTCAC</td>\n",
       "      <td>CCTCCCAACCCCGCACTCACAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>20</td>\n",
       "      <td>0.034682</td>\n",
       "      <td>0.067212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C*07:02:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>AACCCCGCACTCACAGGCCC</td>\n",
       "      <td>AACCCCGCACTCACAGGCCCAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>14</td>\n",
       "      <td>0.115109</td>\n",
       "      <td>0.067212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C*07:02:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>GCACTCACAGGCCCAGGTCT</td>\n",
       "      <td>GCACTCACAGGCCCAGGTCTCGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>8</td>\n",
       "      <td>0.471516</td>\n",
       "      <td>0.067212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C*07:02:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>CACAGGCCCAGGTCTCGGTC</td>\n",
       "      <td>CACAGGCCCAGGTCTCGGTCAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>3</td>\n",
       "      <td>0.341184</td>\n",
       "      <td>0.067212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C*07:02:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>ACAGGCCCAGGTCTCGGTCA</td>\n",
       "      <td>ACAGGCCCAGGTCTCGGTCAGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>2</td>\n",
       "      <td>0.261074</td>\n",
       "      <td>0.067212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1949</th>\n",
       "      <td>C*04:06:01</td>\n",
       "      <td>7</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>CCCTGGGCACTGTTGCTGGC</td>\n",
       "      <td>CCCTGGGCACTGTTGCTGGCTGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>20</td>\n",
       "      <td>0.006267</td>\n",
       "      <td>0.996367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1950</th>\n",
       "      <td>C*04:06:01</td>\n",
       "      <td>7</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>CCTGGGCACTGTTGCTGGCT</td>\n",
       "      <td>CCTGGGCACTGTTGCTGGCTGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>19</td>\n",
       "      <td>0.008453</td>\n",
       "      <td>0.996367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1951</th>\n",
       "      <td>C*04:06:01</td>\n",
       "      <td>7</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>CTGGGCACTGTTGCTGGCTG</td>\n",
       "      <td>CTGGGCACTGTTGCTGGCTGGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>18</td>\n",
       "      <td>0.012549</td>\n",
       "      <td>0.996367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1952</th>\n",
       "      <td>C*04:06:01</td>\n",
       "      <td>7</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>CCCAGCCAGCAACAGTGCCC</td>\n",
       "      <td>CCCAGCCAGCAACAGTGCCCAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>4</td>\n",
       "      <td>0.188218</td>\n",
       "      <td>0.996367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1953</th>\n",
       "      <td>C*04:06:01</td>\n",
       "      <td>7</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>CCAGCCAGCAACAGTGCCCA</td>\n",
       "      <td>CCAGCCAGCAACAGTGCCCAGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>3</td>\n",
       "      <td>0.086171</td>\n",
       "      <td>0.996367</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1954 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     haplotype name exon number splice site type           protospacer  \\\n",
       "0     C*07:02:01:01           1            donor  CCTCCCAACCCCGCACTCAC   \n",
       "1     C*07:02:01:01           1            donor  AACCCCGCACTCACAGGCCC   \n",
       "2     C*07:02:01:01           1            donor  GCACTCACAGGCCCAGGTCT   \n",
       "3     C*07:02:01:01           1            donor  CACAGGCCCAGGTCTCGGTC   \n",
       "4     C*07:02:01:01           1            donor  ACAGGCCCAGGTCTCGGTCA   \n",
       "...             ...         ...              ...                   ...   \n",
       "1949     C*04:06:01           7         acceptor  CCCTGGGCACTGTTGCTGGC   \n",
       "1950     C*04:06:01           7         acceptor  CCTGGGCACTGTTGCTGGCT   \n",
       "1951     C*04:06:01           7         acceptor  CTGGGCACTGTTGCTGGCTG   \n",
       "1952     C*04:06:01           7         acceptor  CCCAGCCAGCAACAGTGCCC   \n",
       "1953     C*04:06:01           7         acceptor  CCAGCCAGCAACAGTGCCCA   \n",
       "\n",
       "           guide seq with PAM  PAM enzyme class CBE/ABE position  guide score  \\\n",
       "0     CCTCCCAACCCCGCACTCACAGG  NGG          CBE               20     0.034682   \n",
       "1     AACCCCGCACTCACAGGCCCAGG  NGG          CBE               14     0.115109   \n",
       "2     GCACTCACAGGCCCAGGTCTCGG  NGG          CBE                8     0.471516   \n",
       "3     CACAGGCCCAGGTCTCGGTCAGG  NGG          CBE                3     0.341184   \n",
       "4     ACAGGCCCAGGTCTCGGTCAGGG  NGG          CBE                2     0.261074   \n",
       "...                       ...  ...          ...              ...          ...   \n",
       "1949  CCCTGGGCACTGTTGCTGGCTGG  NGG          CBE               20     0.006267   \n",
       "1950  CCTGGGCACTGTTGCTGGCTGGG  NGG          CBE               19     0.008453   \n",
       "1951  CTGGGCACTGTTGCTGGCTGGGG  NGG          CBE               18     0.012549   \n",
       "1952  CCCAGCCAGCAACAGTGCCCAGG  NGG          ABE                4     0.188218   \n",
       "1953  CCAGCCAGCAACAGTGCCCAGGG  NGG          ABE                3     0.086171   \n",
       "\n",
       "      cDNA distuption pct  \n",
       "0                0.067212  \n",
       "1                0.067212  \n",
       "2                0.067212  \n",
       "3                0.067212  \n",
       "4                0.067212  \n",
       "...                   ...  \n",
       "1949             0.996367  \n",
       "1950             0.996367  \n",
       "1951             0.996367  \n",
       "1952             0.996367  \n",
       "1953             0.996367  \n",
       "\n",
       "[1954 rows x 10 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#### running spliceR on all HLA C common haplotypes \n",
    "\n",
    "# initializing empty dataframe to store results \n",
    "cols = ['haplotype name', 'exon number', 'splice site type', 'protospacer', 'guide seq with PAM', 'PAM', 'enzyme class', 'CBE/ABE position', 'guide score', 'cDNA disruption pct']\n",
    "HLA_C_output_df = pd.DataFrame(columns=cols)\n",
    "\n",
    "motif_weights_filepath = \"./motif_weights.tsv\"\n",
    "position_weights_filepath = \"./position_weights.tsv\"\n",
    "motif_weights_dict, position_weights_dict = parse_motif_position_weights(motif_weights_filepath, position_weights_filepath)\n",
    "\n",
    "for haplotype in HLA_C_name_to_accession:\n",
    "    print(haplotype)\n",
    "    accession = HLA_C_name_to_accession[haplotype]\n",
    "    hap_url = f\"https://www.ebi.ac.uk/cgi-bin/ipd/api/allele/{accession}\"\n",
    "    # print(hp_url)\n",
    "    hap_info = parse_json(file_url=hap_url)\n",
    "    genomic_features = hap_info['genomic_features']\n",
    "    coding_features = hap_info['coding_features']\n",
    "    genomic_sequence = hap_info['genomic_sequence']\n",
    "    coding_sequence = hap_info['sequences']['coding']\n",
    "\n",
    "    hap_out = spliceR_main(\n",
    "        genomic_features,\n",
    "        coding_features,\n",
    "        genomic_sequence,\n",
    "        coding_sequence,\n",
    "        motif_weights_dict,\n",
    "        position_weights_dict,\n",
    "        output_as_df = True,\n",
    "        haplotype_name = haplotype,\n",
    "        guide_length = 20, \n",
    "        pam = \"NGG\",\n",
    "        flank_length = 20,\n",
    "        CBE = True, \n",
    "        ABE = True,\n",
    "        SA = True, \n",
    "        SD = True,\n",
    "        logistic_adjust = 1\n",
    "    )\n",
    "    HLA_C_output_df = pd.concat([HLA_C_output_df, hap_out], axis=0, ignore_index=True)\n",
    "\n",
    "\n",
    "HLA_C_output_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## saving HLA C raw outputs \n",
    "HLA_C_output_df.to_csv('HLA_C_all_haps_splice_site_BE.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### HLA-DQB1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'DQB1*03:02:01:01': 'HLA00627',\n",
       " 'DQB1*02:01:01:01': 'HLA00622',\n",
       " 'DQB1*05:02:01:01': 'HLA00639',\n",
       " 'DQB1*03:01:01:01': 'HLA00625',\n",
       " 'DQB1*05:03:01:01': 'HLA00640',\n",
       " 'DQB1*06:04:01:01': 'HLA00648',\n",
       " 'DQB1*04:02:01:01': 'HLA00637',\n",
       " 'DQB1*06:09:01:01': 'HLA00654',\n",
       " 'DQB1*06:02:01:01': 'HLA00646',\n",
       " 'DQB1*03:03:02:01': 'HLA00629',\n",
       " 'DQB1*04:01:01:01': 'HLA00636',\n",
       " 'DQB1*06:01:01:01': 'HLA00643',\n",
       " 'DQB1*05:01:01:01': 'HLA00638',\n",
       " 'DQB1*06:03:01:01': 'HLA00647'}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hla_dqb1_n_to_a = pd.read_csv(\"HLA_name_to_dbAccession/HLA_DQB1_name_to_accession.txt\", sep=\"\\t\")\n",
    "hla_dqb1_n_to_a\n",
    "HLA_DQB1_name_to_accession = dict(zip(hla_dqb1_n_to_a['name'], hla_dqb1_n_to_a['accession']))\n",
    "HLA_DQB1_name_to_accession\n",
    "# len(HLA_DQB1_name_to_accession) # 14 target haplotypes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### running spliceR on all HLA DQB1 common haplotypes \n",
    "\n",
    "# initializing empty dataframe to store results \n",
    "cols = ['haplotype name', 'exon number', 'splice site type', 'protospacer', 'guide seq with PAM', 'PAM', 'enzyme class', 'CBE/ABE position', 'guide score', 'cDNA disruption pct']\n",
    "HLA_DQB1_output_df = pd.DataFrame(columns=cols)\n",
    "\n",
    "motif_weights_filepath = \"./motif_weights.tsv\"\n",
    "position_weights_filepath = \"./position_weights.tsv\"\n",
    "motif_weights_dict, position_weights_dict = parse_motif_position_weights(motif_weights_filepath, position_weights_filepath)\n",
    "\n",
    "for haplotype in HLA_DQB1_name_to_accession:\n",
    "    print(haplotype)\n",
    "    accession = HLA_DQB1_name_to_accession[haplotype]\n",
    "    hap_url = f\"https://www.ebi.ac.uk/cgi-bin/ipd/api/allele/{accession}\"\n",
    "    hap_info = parse_json(file_url=hap_url)\n",
    "    genomic_features = hap_info['genomic_features']\n",
    "    coding_features = hap_info['coding_features']\n",
    "    genomic_sequence = hap_info['genomic_sequence']\n",
    "    coding_sequence = hap_info['sequences']['coding']\n",
    "\n",
    "    print(genomic_features)\n",
    "\n",
    "    hap_out = spliceR_main(\n",
    "        genomic_features,\n",
    "        coding_features,\n",
    "        genomic_sequence,\n",
    "        coding_sequence,\n",
    "        motif_weights_dict,\n",
    "        position_weights_dict,\n",
    "        output_as_df = True,\n",
    "        haplotype_name = haplotype,\n",
    "        guide_length = 20, \n",
    "        pam = \"NGG\",\n",
    "        flank_length = 20,\n",
    "        CBE = True, \n",
    "        ABE = True,\n",
    "        SA = True, \n",
    "        SD = True,\n",
    "        logistic_adjust = 1, \n",
    "        max_exon_targeted = 4 #HLA-DQB1 has exon 5 splice acceptor site polymorphisms which causes issues with the code since the motif is not the typical SA site motif (exons 5, 6 are intracellular anyway so are not good targets)\n",
    "    )\n",
    "    HLA_DQB1_output_df = pd.concat([HLA_DQB1_output_df, hap_out], axis=0, ignore_index=True)\n",
    "\n",
    "\n",
    "HLA_DQB1_output_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "## saving HLA DQB1 raw outputs \n",
    "HLA_DQB1_output_df.to_csv('HLA_DQB1_all_haps_splice_site_BE_exclude_exons5_6.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### HLA-DRB1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'DRB1*14:05:01:01': 'HLA00837',\n",
       " 'DRB1*14:04:01:01': 'HLA00836',\n",
       " 'DRB1*15:02:02:01': 'HLA00868',\n",
       " 'DRB1*04:05:01:01': 'HLA00690',\n",
       " 'DRB1*08:02:01:01': 'HLA00724',\n",
       " 'DRB1*04:07:01:01': 'HLA00693',\n",
       " 'DRB1*15:06:01': 'HLA00873',\n",
       " 'DRB1*13:02:01:01': 'HLA00798',\n",
       " 'DRB1*09:01:02:01': 'HLA00749',\n",
       " 'DRB1*08:04:01:01': 'HLA00728',\n",
       " 'DRB1*08:04:04': 'HLA01414',\n",
       " 'DRB1*11:04:01:01': 'HLA00756',\n",
       " 'DRB1*11:04:02': 'HLA00757',\n",
       " 'DRB1*04:06:01': 'HLA00692',\n",
       " 'DRB1*15:03:01:01': 'HLA00870',\n",
       " 'DRB1*13:03:01:01': 'HLA00799',\n",
       " 'DRB1*04:02:01': 'HLA00687',\n",
       " 'DRB1*12:01:01:01': 'HLA00789',\n",
       " 'DRB1*04:01:01:01': 'HLA00685',\n",
       " 'DRB1*01:02:01:01': 'HLA00665',\n",
       " 'DRB1*11:01:01:01': 'HLA00751',\n",
       " 'DRB1*11:01:02:01': 'HLA00752',\n",
       " 'DRB1*14:02:01:01': 'HLA00834',\n",
       " 'DRB1*15:01:01:01': 'HLA00865',\n",
       " 'DRB1*03:02:01:01': 'HLA00673',\n",
       " 'DRB1*14:01:01:01': 'HLA00833',\n",
       " 'DRB1*03:01:01:01': 'HLA00671',\n",
       " 'DRB1*01:01:01:01': 'HLA00664',\n",
       " 'DRB1*01:01:02': 'HLA01598',\n",
       " 'DRB1*12:02:01:01': 'HLA00790',\n",
       " 'DRB1*12:02:02:01': 'HLA00791',\n",
       " 'DRB1*14:03:01': 'HLA00835',\n",
       " 'DRB1*13:04': 'HLA00801',\n",
       " 'DRB1*08:01:01:01': 'HLA00723',\n",
       " 'DRB1*04:04:01:01': 'HLA00689',\n",
       " 'DRB1*14:06:01': 'HLA00838',\n",
       " 'DRB1*13:01:01:01': 'HLA00797',\n",
       " 'DRB1*13:12:01': 'HLA00810',\n",
       " 'DRB1*11:02:01:01': 'HLA00754',\n",
       " 'DRB1*01:03:01:01': 'HLA00667',\n",
       " 'DRB1*16:01:01:01': 'HLA00876',\n",
       " 'DRB1*04:11:01:01': 'HLA00697',\n",
       " 'DRB1*04:10:01': 'HLA00696',\n",
       " 'DRB1*07:01:01:01': 'HLA00719',\n",
       " 'DRB1*10:01:01:01': 'HLA00750',\n",
       " 'DRB1*04:03:01:01': 'HLA00688',\n",
       " 'DRB1*16:02:01:01': 'HLA00878',\n",
       " 'DRB1*08:03:02:01': 'HLA00727',\n",
       " 'DRB1*14:08': 'HLA00840'}"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hla_drb1_n_to_a = pd.read_csv(\"HLA_name_to_dbAccession/HLA_DRB1_name_to_accession.txt\", sep=\"\\t\")\n",
    "hla_drb1_n_to_a\n",
    "HLA_DRB1_name_to_accession = dict(zip(hla_drb1_n_to_a['name'], hla_drb1_n_to_a['accession']))\n",
    "HLA_DRB1_name_to_accession"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DRB1*14:05:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*14:04:01:01\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/sm/2bd25dyd1bx5nl6h9b6j4z7c0000gn/T/ipykernel_14360/1043908064.py:40: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  HLA_DRB1_output_df = pd.concat([HLA_DRB1_output_df, hap_out], axis=0, ignore_index=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*15:02:02:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*04:05:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*08:02:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*04:07:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*15:06:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*13:02:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*09:01:02:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*08:04:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*08:04:04\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*11:04:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*11:04:02\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*04:06:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*15:03:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*13:03:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*04:02:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*12:01:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*04:01:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*01:02:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*11:01:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*11:01:02:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*14:02:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*15:01:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*03:02:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*14:01:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*03:01:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*01:01:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*01:01:02\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*12:02:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*12:02:02:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*14:03:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*13:04\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*08:01:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*04:04:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*14:06:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*13:01:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*13:12:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*11:02:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*01:03:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*16:01:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*04:11:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*04:10:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*07:01:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*10:01:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*04:03:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*16:02:01:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*08:03:02:01\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "DRB1*14:08\n",
      "num exons 6\n",
      "exon 1\n",
      "exon 2\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 3\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n",
      "exon 4\n",
      "SA, CBE\n",
      "SA, ABE\n",
      "SD, CBE\n",
      "SD, ABE\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>haplotype name</th>\n",
       "      <th>exon number</th>\n",
       "      <th>splice site type</th>\n",
       "      <th>protospacer</th>\n",
       "      <th>guide seq with PAM</th>\n",
       "      <th>PAM</th>\n",
       "      <th>enzyme class</th>\n",
       "      <th>CBE/ABE position</th>\n",
       "      <th>guide score</th>\n",
       "      <th>cDNA disruption pct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DRB1*14:05:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>CCACAATGTGCACTTACGTC</td>\n",
       "      <td>CCACAATGTGCACTTACGTCTGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>17</td>\n",
       "      <td>0.048857</td>\n",
       "      <td>0.126092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DRB1*14:05:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>CCACAATGTGCACTTACGTC</td>\n",
       "      <td>CCACAATGTGCACTTACGTCTGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>16</td>\n",
       "      <td>0.045911</td>\n",
       "      <td>0.126092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DRB1*14:05:01:01</td>\n",
       "      <td>2</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>GTACTCCAAGAAACGTGCTG</td>\n",
       "      <td>GTACTCCAAGAAACGTGCTGTGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>18</td>\n",
       "      <td>0.012222</td>\n",
       "      <td>0.463171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DRB1*14:05:01:01</td>\n",
       "      <td>2</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>TACTCCAAGAAACGTGCTGT</td>\n",
       "      <td>TACTCCAAGAAACGTGCTGTGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>17</td>\n",
       "      <td>0.020726</td>\n",
       "      <td>0.463171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DRB1*14:05:01:01</td>\n",
       "      <td>2</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>ACTCCAAGAAACGTGCTGTG</td>\n",
       "      <td>ACTCCAAGAAACGTGCTGTGGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>16</td>\n",
       "      <td>0.024067</td>\n",
       "      <td>0.463171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>951</th>\n",
       "      <td>DRB1*14:08</td>\n",
       "      <td>3</td>\n",
       "      <td>donor</td>\n",
       "      <td>TCACTCCATTCCACTGTGAG</td>\n",
       "      <td>TCACTCCATTCCACTGTGAGAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>3</td>\n",
       "      <td>0.144056</td>\n",
       "      <td>0.815231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>952</th>\n",
       "      <td>DRB1*14:08</td>\n",
       "      <td>3</td>\n",
       "      <td>donor</td>\n",
       "      <td>CACTCCATTCCACTGTGAGA</td>\n",
       "      <td>CACTCCATTCCACTGTGAGAGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>2</td>\n",
       "      <td>0.075413</td>\n",
       "      <td>0.815231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>953</th>\n",
       "      <td>DRB1*14:08</td>\n",
       "      <td>4</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>TCAGACCGTGCTCCTGAGAG</td>\n",
       "      <td>TCAGACCGTGCTCCTGAGAGAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>14</td>\n",
       "      <td>0.116474</td>\n",
       "      <td>0.953808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>954</th>\n",
       "      <td>DRB1*14:08</td>\n",
       "      <td>4</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>TGCTCCTGAGAGAGGAAGCC</td>\n",
       "      <td>TGCTCCTGAGAGAGGAAGCCAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>6</td>\n",
       "      <td>0.843209</td>\n",
       "      <td>0.953808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>955</th>\n",
       "      <td>DRB1*14:08</td>\n",
       "      <td>4</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>GGCTTCCTCTCTCAGGAGCA</td>\n",
       "      <td>GGCTTCCTCTCTCAGGAGCACGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>14</td>\n",
       "      <td>0.022565</td>\n",
       "      <td>0.953808</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>956 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       haplotype name exon number splice site type           protospacer  \\\n",
       "0    DRB1*14:05:01:01           1            donor  CCACAATGTGCACTTACGTC   \n",
       "1    DRB1*14:05:01:01           1            donor  CCACAATGTGCACTTACGTC   \n",
       "2    DRB1*14:05:01:01           2         acceptor  GTACTCCAAGAAACGTGCTG   \n",
       "3    DRB1*14:05:01:01           2         acceptor  TACTCCAAGAAACGTGCTGT   \n",
       "4    DRB1*14:05:01:01           2         acceptor  ACTCCAAGAAACGTGCTGTG   \n",
       "..                ...         ...              ...                   ...   \n",
       "951        DRB1*14:08           3            donor  TCACTCCATTCCACTGTGAG   \n",
       "952        DRB1*14:08           3            donor  CACTCCATTCCACTGTGAGA   \n",
       "953        DRB1*14:08           4         acceptor  TCAGACCGTGCTCCTGAGAG   \n",
       "954        DRB1*14:08           4         acceptor  TGCTCCTGAGAGAGGAAGCC   \n",
       "955        DRB1*14:08           4         acceptor  GGCTTCCTCTCTCAGGAGCA   \n",
       "\n",
       "          guide seq with PAM  PAM enzyme class CBE/ABE position  guide score  \\\n",
       "0    CCACAATGTGCACTTACGTCTGG  NGG          CBE               17     0.048857   \n",
       "1    CCACAATGTGCACTTACGTCTGG  NGG          ABE               16     0.045911   \n",
       "2    GTACTCCAAGAAACGTGCTGTGG  NGG          CBE               18     0.012222   \n",
       "3    TACTCCAAGAAACGTGCTGTGGG  NGG          CBE               17     0.020726   \n",
       "4    ACTCCAAGAAACGTGCTGTGGGG  NGG          CBE               16     0.024067   \n",
       "..                       ...  ...          ...              ...          ...   \n",
       "951  TCACTCCATTCCACTGTGAGAGG  NGG          ABE                3     0.144056   \n",
       "952  CACTCCATTCCACTGTGAGAGGG  NGG          ABE                2     0.075413   \n",
       "953  TCAGACCGTGCTCCTGAGAGAGG  NGG          CBE               14     0.116474   \n",
       "954  TGCTCCTGAGAGAGGAAGCCAGG  NGG          CBE                6     0.843209   \n",
       "955  GGCTTCCTCTCTCAGGAGCACGG  NGG          ABE               14     0.022565   \n",
       "\n",
       "     cDNA disruption pct  \n",
       "0               0.126092  \n",
       "1               0.126092  \n",
       "2               0.463171  \n",
       "3               0.463171  \n",
       "4               0.463171  \n",
       "..                   ...  \n",
       "951             0.815231  \n",
       "952             0.815231  \n",
       "953             0.953808  \n",
       "954             0.953808  \n",
       "955             0.953808  \n",
       "\n",
       "[956 rows x 10 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#### running spliceR on all HLA DRB1 common haplotypes \n",
    "\n",
    "# initializing empty dataframe to store results \n",
    "cols = ['haplotype name', 'exon number', 'splice site type', 'protospacer', 'guide seq with PAM', 'PAM', 'enzyme class', 'CBE/ABE position', 'guide score', 'cDNA disruption pct']\n",
    "HLA_DRB1_output_df = pd.DataFrame(columns=cols)\n",
    "\n",
    "motif_weights_filepath = \"./motif_weights.tsv\"\n",
    "position_weights_filepath = \"./position_weights.tsv\"\n",
    "motif_weights_dict, position_weights_dict = parse_motif_position_weights(motif_weights_filepath, position_weights_filepath)\n",
    "\n",
    "for haplotype in HLA_DRB1_name_to_accession:\n",
    "    print(haplotype)\n",
    "    accession = HLA_DRB1_name_to_accession[haplotype]\n",
    "    hap_url = f\"https://www.ebi.ac.uk/cgi-bin/ipd/api/allele/{accession}\"\n",
    "    hap_info = parse_json(file_url=hap_url)\n",
    "    genomic_features = hap_info['genomic_features']\n",
    "    coding_features = hap_info['coding_features']\n",
    "    genomic_sequence = hap_info['genomic_sequence']\n",
    "    coding_sequence = hap_info['sequences']['coding']\n",
    "\n",
    "    hap_out = spliceR_main(\n",
    "        genomic_features,\n",
    "        coding_features,\n",
    "        genomic_sequence,\n",
    "        coding_sequence,\n",
    "        motif_weights_dict,\n",
    "        position_weights_dict,\n",
    "        output_as_df = True,\n",
    "        haplotype_name = haplotype,\n",
    "        guide_length = 20, \n",
    "        pam = \"NGG\",\n",
    "        flank_length = 20,\n",
    "        CBE = True, \n",
    "        ABE = True,\n",
    "        SA = True, \n",
    "        SD = True,\n",
    "        logistic_adjust = 1, \n",
    "        max_exon_targeted = 4 #exons 5, 6 are intracellular so are not good targets\n",
    "    )\n",
    "    HLA_DRB1_output_df = pd.concat([HLA_DRB1_output_df, hap_out], axis=0, ignore_index=True)\n",
    "\n",
    "\n",
    "HLA_DRB1_output_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "## saving HLA DQB1 raw outputs \n",
    "HLA_DRB1_output_df.to_csv('HLA_DRB1_all_haps_splice_site_BE_exclude_exons5_6.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### debugging"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "exons list [{'length': 109, 'number': '1', 'partial': False, 'start': 530, 'type': 'exon'}, {'length': 270, 'number': '2', 'partial': False, 'start': 2083, 'type': 'exon'}, {'length': 282, 'number': '3', 'partial': False, 'start': 4883, 'type': 'exon'}, {'length': 111, 'number': '4', 'partial': False, 'start': 5680, 'type': 'exon'}, {'length': 24, 'number': '5', 'partial': False, 'start': 6273, 'type': 'exon'}, {'length': 14, 'number': '6', 'partial': False, 'start': 6907, 'type': 'exon'}]\n",
      "exon 1\n",
      "exon start 529\n",
      "start of exon ATGTC\n",
      "sa_start 519\n",
      "sa_end 539\n",
      "sa_seq TCCAAGACATAATTGAGACG\n",
      "motif CATAATT\n",
      "exon 2\n",
      "exon start 2082\n",
      "start of exon AGGAT\n",
      "sa_start 2072\n",
      "sa_end 2092\n",
      "sa_seq ACGAAATCCTCTGCGGGGAA\n",
      "motif CCTCTGC\n",
      "exon 3\n",
      "exon start 4882\n",
      "start of exon TGGAG\n",
      "sa_start 4872\n",
      "sa_end 4892\n",
      "sa_seq GTGGGCTCCACTGAGGGCAG\n",
      "motif CCACTGA\n",
      "exon 4\n",
      "exon start 5679\n",
      "start of exon GGGCT\n",
      "sa_start 5669\n",
      "sa_end 5689\n",
      "sa_seq GACTGAGCCCCTAAGAAGCA\n",
      "motif CCCCTAA\n",
      "exon 5\n",
      "exon start 6272\n",
      "start of exon GACCT\n",
      "sa_start 6262\n",
      "sa_end 6282\n",
      "sa_seq CCTTGAGGTCTTACAAAAGG\n",
      "motif GTCTTAC\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'GTCTTAC'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyError\u001b[39m                                  Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[30]\u001b[39m\u001b[32m, line 58\u001b[39m\n\u001b[32m     56\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[32m1\u001b[39m<=target_pos<=\u001b[32m20\u001b[39m:\n\u001b[32m     57\u001b[39m     pw = position_weights_dict[\u001b[33m\"\u001b[39m\u001b[33mCBE\u001b[39m\u001b[33m\"\u001b[39m][target_pos][\u001b[33m'\u001b[39m\u001b[33mposition_weight\u001b[39m\u001b[33m'\u001b[39m]\n\u001b[32m---> \u001b[39m\u001b[32m58\u001b[39m     mw = \u001b[43mmotif_weights_dict\u001b[49m\u001b[43m[\u001b[49m\u001b[43mmotif\u001b[49m\u001b[43m]\u001b[49m[\u001b[33m'\u001b[39m\u001b[33mmotif_weight\u001b[39m\u001b[33m'\u001b[39m]\n\u001b[32m     59\u001b[39m     cbe_score = probability(mw+pw-logistic_adjust)\n\u001b[32m     60\u001b[39m     u_cbe_sd_guide_wo_pam = u_cbe_sd_guide[\u001b[32m1\u001b[39m][:-\u001b[32m3\u001b[39m]\n",
      "\u001b[31mKeyError\u001b[39m: 'GTCTTAC'"
     ]
    }
   ],
   "source": [
    "exons_list = []\n",
    "num_exons = 0 \n",
    "for feature in genomic_features:\n",
    "    if feature['type'] =='exon':\n",
    "        exons_list.append(feature)\n",
    "        if int(feature['number'])>num_exons:\n",
    "            num_exons=int(feature['number'])\n",
    "\n",
    "exons_list\n",
    "print(\"exons list\", exons_list)\n",
    "\n",
    "pam=\"NGG\"\n",
    "flank_length=7\n",
    "guide_length=5\n",
    "haplotype_name= \"DQB1*02:01:01:01\"\n",
    "logistic_adjust = 1\n",
    "pct_of_cds = 0\n",
    "\n",
    "for i, exon in enumerate(exons_list):\n",
    "    exon_number = i\n",
    "    print(\"exon\", exon['number'])\n",
    "\n",
    "    ## calculating what percentage of the way through coding sequence the end of this exon is (want earlier disruption)\n",
    "    # end_of_exon_in_cds = 0 \n",
    "    # for cd_feat in coding_features:\n",
    "    #     if cd_feat['number'] == exon['number']:\n",
    "    #         end_of_exon_in_cds=cd_feat['start']+cd_feat['length']\n",
    "    #         break\n",
    "    # pct_of_cds = end_of_exon_in_cds/len(coding_sequence)\n",
    "    \n",
    "    ## getting start and end coordinates of this exon (relative to GENOMIC sequence)\n",
    "    exon_start_idx = exon['start']-1 #-1 to account for 0 indexing \n",
    "    exon_end_idx = exon_start_idx+exon['length']\n",
    "    \n",
    "    print(\"exon start\", exon_start_idx)\n",
    "    print(\"start of exon\", genomic_sequence[exon_start_idx:exon_start_idx+5])\n",
    "    pam_length = len(pam)\n",
    "    sa_start = exon_start_idx-flank_length-pam_length\n",
    "    sa_end = exon_start_idx+flank_length+pam_length\n",
    "\n",
    "    print(\"sa_start\", sa_start)\n",
    "    print(\"sa_end\", sa_end)\n",
    "\n",
    "    SA_seq = get_rev_complement(genomic_sequence[sa_start:sa_end])\n",
    "    guide_template_w_pam = \"N\"*guide_length+pam\n",
    "    unscored_CBE_SA_guides = matchDNAPattern(guide_template_w_pam, SA_seq)\n",
    "    motif = get_rev_complement(genomic_sequence[exon_start_idx-4:exon_start_idx+3]) #checked \n",
    "\n",
    "    print(\"sa_seq\", SA_seq)\n",
    "    print(\"motif\", motif)\n",
    "\n",
    "    output = []\n",
    "    for u_cbe_sd_guide in unscored_CBE_SA_guides:\n",
    "        strt = u_cbe_sd_guide[0]\n",
    "        target_pos = flank_length+len(pam)-strt #position of target base within the guide (1 to 20 + pam)\n",
    "        if 1<=target_pos<=20:\n",
    "            pw = position_weights_dict[\"CBE\"][target_pos]['position_weight']\n",
    "            mw = motif_weights_dict[motif]['motif_weight']\n",
    "            cbe_score = probability(mw+pw-logistic_adjust)\n",
    "            u_cbe_sd_guide_wo_pam = u_cbe_sd_guide[1][:-3]\n",
    "            guide_info = [haplotype_name, exon_number, \"donor\",u_cbe_sd_guide_wo_pam, u_cbe_sd_guide[1], pam, \"CBE\", target_pos, cbe_score, pct_of_cds]\n",
    "            output.append(guide_info)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_CBE_SA_guides(haplotype_name, exon_number, pam, exon_start_idx, flank_length, genomic_sequence, position_weights_dict, motif_weights_dict, guide_length, pct_of_cds):\n",
    "    pam_length = len(pam)\n",
    "    sa_start = exon_start_idx-flank_length-pam_length\n",
    "    sa_end = exon_start_idx+flank_length+pam_length\n",
    "    SA_seq = get_rev_complement(genomic_sequence[sa_start:sa_end])\n",
    "    guide_template_w_pam = \"N\"*guide_length+pam\n",
    "    unscored_CBE_SA_guides = matchDNAPattern(guide_template_w_pam, SA_seq)\n",
    "    motif = get_rev_complement(genomic_sequence[exon_start_idx-4:exon_start_idx+3]) #checked \n",
    "    # scoring all potential guides\n",
    "    output = []\n",
    "    for u_cbe_sa_guide in unscored_CBE_SA_guides: \n",
    "        strt = u_cbe_sa_guide[0]\n",
    "        target_pos = flank_length+len(pam)-strt+1 #position of target base within the guide (1 to 20 + pam) checked\n",
    "        if 1<=target_pos<=20:\n",
    "            pw = position_weights_dict[\"CBE\"][target_pos]['position_weight']\n",
    "            mw = motif_weights_dict[motif]['motif_weight']\n",
    "            cbe_score = probability(mw+pw-logistic_adjust)\n",
    "            u_cbe_sa_guide_wo_pam = u_cbe_sa_guide[1][:-3]\n",
    "            guide_info = [haplotype_name, exon_number, \"acceptor\", u_cbe_sa_guide_wo_pam, u_cbe_sa_guide[1], pam, \"CBE\", target_pos, cbe_score, pct_of_cds]\n",
    "            output.append(guide_info)\n",
    "    return output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'GTCTTAC'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyError\u001b[39m                                  Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[10]\u001b[39m\u001b[32m, line 2\u001b[39m\n\u001b[32m      1\u001b[39m motif_weights_dict, position_weights_dict = parse_motif_position_weights(motif_weights_filepath, position_weights_filepath)\n\u001b[32m----> \u001b[39m\u001b[32m2\u001b[39m \u001b[43mmotif_weights_dict\u001b[49m\u001b[43m[\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mGTCTTAC\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m]\u001b[49m\n",
      "\u001b[31mKeyError\u001b[39m: 'GTCTTAC'"
     ]
    }
   ],
   "source": [
    "motif_weights_dict, position_weights_dict = parse_motif_position_weights(motif_weights_filepath, position_weights_filepath)\n",
    "motif_weights_dict['GTCTTAC']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## tests/extras"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>haplotype name</th>\n",
       "      <th>exon number</th>\n",
       "      <th>splice site type</th>\n",
       "      <th>protospacer</th>\n",
       "      <th>guide seq with PAM</th>\n",
       "      <th>PAM</th>\n",
       "      <th>enzyme class</th>\n",
       "      <th>CBE/ABE position</th>\n",
       "      <th>guide score</th>\n",
       "      <th>cDNA distuption pct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>GACCCCGCACTCACCTGCCC</td>\n",
       "      <td>GACCCCGCACTCACCTGCCCAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>14</td>\n",
       "      <td>0.115790</td>\n",
       "      <td>0.067395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>CGCACTCACCTGCCCAGGTC</td>\n",
       "      <td>CGCACTCACCTGCCCAGGTCTGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>9</td>\n",
       "      <td>0.340876</td>\n",
       "      <td>0.067395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>GCACTCACCTGCCCAGGTCT</td>\n",
       "      <td>GCACTCACCTGCCCAGGTCTGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>8</td>\n",
       "      <td>0.473178</td>\n",
       "      <td>0.067395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>CACCTGCCCAGGTCTGGGTC</td>\n",
       "      <td>CACCTGCCCAGGTCTGGGTCAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>3</td>\n",
       "      <td>0.342684</td>\n",
       "      <td>0.067395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>ACCTGCCCAGGTCTGGGTCA</td>\n",
       "      <td>ACCTGCCCAGGTCTGGGTCAGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>2</td>\n",
       "      <td>0.262362</td>\n",
       "      <td>0.067395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>GACCCCGCACTCACCTGCCC</td>\n",
       "      <td>GACCCCGCACTCACCTGCCCAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>13</td>\n",
       "      <td>0.020324</td>\n",
       "      <td>0.067395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>CGCACTCACCTGCCCAGGTC</td>\n",
       "      <td>CGCACTCACCTGCCCAGGTCTGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>8</td>\n",
       "      <td>0.269631</td>\n",
       "      <td>0.067395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>GCACTCACCTGCCCAGGTCT</td>\n",
       "      <td>GCACTCACCTGCCCAGGTCTGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>7</td>\n",
       "      <td>0.417329</td>\n",
       "      <td>0.067395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>CACCTGCCCAGGTCTGGGTC</td>\n",
       "      <td>CACCTGCCCAGGTCTGGGTCAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>2</td>\n",
       "      <td>0.061594</td>\n",
       "      <td>0.067395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>ACCTGCCCAGGTCTGGGTCA</td>\n",
       "      <td>ACCTGCCCAGGTCTGGGTCAGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>1</td>\n",
       "      <td>0.027279</td>\n",
       "      <td>0.067395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>2</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>TACCTCATGGAGTGGGAGCC</td>\n",
       "      <td>TACCTCATGGAGTGGGAGCCTGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>20</td>\n",
       "      <td>0.042578</td>\n",
       "      <td>0.313297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>2</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>ACCTCATGGAGTGGGAGCCT</td>\n",
       "      <td>ACCTCATGGAGTGGGAGCCTGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>19</td>\n",
       "      <td>0.056709</td>\n",
       "      <td>0.313297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>2</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>CCTCATGGAGTGGGAGCCTG</td>\n",
       "      <td>CCTCATGGAGTGGGAGCCTGGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>18</td>\n",
       "      <td>0.082248</td>\n",
       "      <td>0.313297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>2</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>GGAGTGGGAGCCTGGGGACG</td>\n",
       "      <td>GGAGTGGGAGCCTGGGGACGAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>12</td>\n",
       "      <td>0.207491</td>\n",
       "      <td>0.313297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>2</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>GGGAGCCTGGGGACGAGGAG</td>\n",
       "      <td>GGGAGCCTGGGGACGAGGAGTGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>7</td>\n",
       "      <td>0.670896</td>\n",
       "      <td>0.313297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>2</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>CCCCAGGCTCCCACTCCATG</td>\n",
       "      <td>CCCCAGGCTCCCACTCCATGAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>5</td>\n",
       "      <td>0.454827</td>\n",
       "      <td>0.313297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>2</td>\n",
       "      <td>donor</td>\n",
       "      <td>CCCGGGCCGGGGTCACTCAC</td>\n",
       "      <td>CCCGGGCCGGGGTCACTCACCGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>20</td>\n",
       "      <td>0.044239</td>\n",
       "      <td>0.313297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>2</td>\n",
       "      <td>donor</td>\n",
       "      <td>GTCACTCACCGGCCTCGCTC</td>\n",
       "      <td>GTCACTCACCGGCCTCGCTCTGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>9</td>\n",
       "      <td>0.398258</td>\n",
       "      <td>0.313297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>2</td>\n",
       "      <td>donor</td>\n",
       "      <td>CCCGGGCCGGGGTCACTCAC</td>\n",
       "      <td>CCCGGGCCGGGGTCACTCACCGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>19</td>\n",
       "      <td>0.028549</td>\n",
       "      <td>0.313297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>2</td>\n",
       "      <td>donor</td>\n",
       "      <td>GTCACTCACCGGCCTCGCTC</td>\n",
       "      <td>GTCACTCACCGGCCTCGCTCTGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>8</td>\n",
       "      <td>0.343390</td>\n",
       "      <td>0.313297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>3</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>ATCTGGAGGGTGTGAGAACC</td>\n",
       "      <td>ATCTGGAGGGTGTGAGAACCTGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>20</td>\n",
       "      <td>0.039942</td>\n",
       "      <td>0.564663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>3</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>AGAACCTGGCCCCGACCCCG</td>\n",
       "      <td>AGAACCTGGCCCCGACCCCGCGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>6</td>\n",
       "      <td>0.860042</td>\n",
       "      <td>0.564663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>4</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>TGTGTCTTGGGGGGGTCTGA</td>\n",
       "      <td>TGTGTCTTGGGGGGGTCTGACGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>17</td>\n",
       "      <td>0.171443</td>\n",
       "      <td>0.816029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>4</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>GTGTCTTGGGGGGGTCTGAC</td>\n",
       "      <td>GTGTCTTGGGGGGGTCTGACGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>16</td>\n",
       "      <td>0.194259</td>\n",
       "      <td>0.816029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>4</td>\n",
       "      <td>donor</td>\n",
       "      <td>CTCCCTCCTTACCCCATCTC</td>\n",
       "      <td>CTCCCTCCTTACCCCATCTCAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>12</td>\n",
       "      <td>0.130095</td>\n",
       "      <td>0.816029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>4</td>\n",
       "      <td>donor</td>\n",
       "      <td>TCCCTCCTTACCCCATCTCA</td>\n",
       "      <td>TCCCTCCTTACCCCATCTCAGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>11</td>\n",
       "      <td>0.164163</td>\n",
       "      <td>0.816029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>4</td>\n",
       "      <td>donor</td>\n",
       "      <td>CCTTACCCCATCTCAGGGTG</td>\n",
       "      <td>CCTTACCCCATCTCAGGGTGAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>6</td>\n",
       "      <td>0.789564</td>\n",
       "      <td>0.816029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>4</td>\n",
       "      <td>donor</td>\n",
       "      <td>CTTACCCCATCTCAGGGTGA</td>\n",
       "      <td>CTTACCCCATCTCAGGGTGAGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>5</td>\n",
       "      <td>0.573362</td>\n",
       "      <td>0.816029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>4</td>\n",
       "      <td>donor</td>\n",
       "      <td>TTACCCCATCTCAGGGTGAG</td>\n",
       "      <td>TTACCCCATCTCAGGGTGAGGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>4</td>\n",
       "      <td>0.392107</td>\n",
       "      <td>0.816029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>4</td>\n",
       "      <td>donor</td>\n",
       "      <td>CTCCCTCCTTACCCCATCTC</td>\n",
       "      <td>CTCCCTCCTTACCCCATCTCAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>11</td>\n",
       "      <td>0.129198</td>\n",
       "      <td>0.816029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>4</td>\n",
       "      <td>donor</td>\n",
       "      <td>TCCCTCCTTACCCCATCTCA</td>\n",
       "      <td>TCCCTCCTTACCCCATCTCAGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>10</td>\n",
       "      <td>0.185686</td>\n",
       "      <td>0.816029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>4</td>\n",
       "      <td>donor</td>\n",
       "      <td>CCTTACCCCATCTCAGGGTG</td>\n",
       "      <td>CCTTACCCCATCTCAGGGTGAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>5</td>\n",
       "      <td>0.604024</td>\n",
       "      <td>0.816029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>4</td>\n",
       "      <td>donor</td>\n",
       "      <td>CTTACCCCATCTCAGGGTGA</td>\n",
       "      <td>CTTACCCCATCTCAGGGTGAGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>4</td>\n",
       "      <td>0.363967</td>\n",
       "      <td>0.816029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>4</td>\n",
       "      <td>donor</td>\n",
       "      <td>TTACCCCATCTCAGGGTGAG</td>\n",
       "      <td>TTACCCCATCTCAGGGTGAGGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>3</td>\n",
       "      <td>0.188794</td>\n",
       "      <td>0.816029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>5</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>GTGGGCTGGGAAGATGGCTC</td>\n",
       "      <td>GTGGGCTGGGAAGATGGCTCTGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>20</td>\n",
       "      <td>0.090060</td>\n",
       "      <td>0.922587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>5</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>TGGGCTGGGAAGATGGCTCT</td>\n",
       "      <td>TGGGCTGGGAAGATGGCTCTGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>19</td>\n",
       "      <td>0.118008</td>\n",
       "      <td>0.922587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>5</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>GAAGATGGCTCTGGGAAAAG</td>\n",
       "      <td>GAAGATGGCTCTGGGAAAAGAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>11</td>\n",
       "      <td>0.433502</td>\n",
       "      <td>0.922587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>5</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>AAGATGGCTCTGGGAAAAGA</td>\n",
       "      <td>AAGATGGCTCTGGGAAAAGAGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>10</td>\n",
       "      <td>0.494032</td>\n",
       "      <td>0.922587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>5</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>AGATGGCTCTGGGAAAAGAG</td>\n",
       "      <td>AGATGGCTCTGGGAAAAGAGGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>9</td>\n",
       "      <td>0.585956</td>\n",
       "      <td>0.922587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>5</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>GGCTCTGGGAAAAGAGGGGA</td>\n",
       "      <td>GGCTCTGGGAAAAGAGGGGAAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>5</td>\n",
       "      <td>0.839644</td>\n",
       "      <td>0.922587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>6</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>GCTCCCTCCTTTTCTATCTG</td>\n",
       "      <td>GCTCCCTCCTTTTCTATCTGTGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>18</td>\n",
       "      <td>0.119323</td>\n",
       "      <td>0.952641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>6</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>CTCCCTCCTTTTCTATCTGT</td>\n",
       "      <td>CTCCCTCCTTTTCTATCTGTGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>17</td>\n",
       "      <td>0.188146</td>\n",
       "      <td>0.952641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>6</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>TTTCTTCCCACAGATAGAAA</td>\n",
       "      <td>TTTCTTCCCACAGATAGAAAAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>12</td>\n",
       "      <td>0.035618</td>\n",
       "      <td>0.952641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>6</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>CTTCCCACAGATAGAAAAGG</td>\n",
       "      <td>CTTCCCACAGATAGAAAAGGAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>9</td>\n",
       "      <td>0.112869</td>\n",
       "      <td>0.952641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>6</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>TTCCCACAGATAGAAAAGGA</td>\n",
       "      <td>TTCCCACAGATAGAAAAGGAGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>8</td>\n",
       "      <td>0.187633</td>\n",
       "      <td>0.952641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>7</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>CCCTGGGCACTGTCACTGCC</td>\n",
       "      <td>CCCTGGGCACTGTCACTGCCTGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>20</td>\n",
       "      <td>0.044522</td>\n",
       "      <td>0.996357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>7</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>CCTGGGCACTGTCACTGCCT</td>\n",
       "      <td>CCTGGGCACTGTCACTGCCTGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>19</td>\n",
       "      <td>0.059258</td>\n",
       "      <td>0.996357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>7</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>CTGGGCACTGTCACTGCCTG</td>\n",
       "      <td>CTGGGCACTGTCACTGCCTGGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>CBE</td>\n",
       "      <td>18</td>\n",
       "      <td>0.085840</td>\n",
       "      <td>0.996357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>7</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>CCCAGGCAGTGACAGTGCCC</td>\n",
       "      <td>CCCAGGCAGTGACAGTGCCCAGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>4</td>\n",
       "      <td>0.238371</td>\n",
       "      <td>0.996357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>A*24:02:01:01</td>\n",
       "      <td>7</td>\n",
       "      <td>acceptor</td>\n",
       "      <td>CCAGGCAGTGACAGTGCCCA</td>\n",
       "      <td>CCAGGCAGTGACAGTGCCCAGGG</td>\n",
       "      <td>NGG</td>\n",
       "      <td>ABE</td>\n",
       "      <td>3</td>\n",
       "      <td>0.112914</td>\n",
       "      <td>0.996357</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   haplotype name exon number splice site type           protospacer  \\\n",
       "0   A*24:02:01:01           1            donor  GACCCCGCACTCACCTGCCC   \n",
       "1   A*24:02:01:01           1            donor  CGCACTCACCTGCCCAGGTC   \n",
       "2   A*24:02:01:01           1            donor  GCACTCACCTGCCCAGGTCT   \n",
       "3   A*24:02:01:01           1            donor  CACCTGCCCAGGTCTGGGTC   \n",
       "4   A*24:02:01:01           1            donor  ACCTGCCCAGGTCTGGGTCA   \n",
       "5   A*24:02:01:01           1            donor  GACCCCGCACTCACCTGCCC   \n",
       "6   A*24:02:01:01           1            donor  CGCACTCACCTGCCCAGGTC   \n",
       "7   A*24:02:01:01           1            donor  GCACTCACCTGCCCAGGTCT   \n",
       "8   A*24:02:01:01           1            donor  CACCTGCCCAGGTCTGGGTC   \n",
       "9   A*24:02:01:01           1            donor  ACCTGCCCAGGTCTGGGTCA   \n",
       "10  A*24:02:01:01           2         acceptor  TACCTCATGGAGTGGGAGCC   \n",
       "11  A*24:02:01:01           2         acceptor  ACCTCATGGAGTGGGAGCCT   \n",
       "12  A*24:02:01:01           2         acceptor  CCTCATGGAGTGGGAGCCTG   \n",
       "13  A*24:02:01:01           2         acceptor  GGAGTGGGAGCCTGGGGACG   \n",
       "14  A*24:02:01:01           2         acceptor  GGGAGCCTGGGGACGAGGAG   \n",
       "15  A*24:02:01:01           2         acceptor  CCCCAGGCTCCCACTCCATG   \n",
       "16  A*24:02:01:01           2            donor  CCCGGGCCGGGGTCACTCAC   \n",
       "17  A*24:02:01:01           2            donor  GTCACTCACCGGCCTCGCTC   \n",
       "18  A*24:02:01:01           2            donor  CCCGGGCCGGGGTCACTCAC   \n",
       "19  A*24:02:01:01           2            donor  GTCACTCACCGGCCTCGCTC   \n",
       "20  A*24:02:01:01           3         acceptor  ATCTGGAGGGTGTGAGAACC   \n",
       "21  A*24:02:01:01           3         acceptor  AGAACCTGGCCCCGACCCCG   \n",
       "22  A*24:02:01:01           4         acceptor  TGTGTCTTGGGGGGGTCTGA   \n",
       "23  A*24:02:01:01           4         acceptor  GTGTCTTGGGGGGGTCTGAC   \n",
       "24  A*24:02:01:01           4            donor  CTCCCTCCTTACCCCATCTC   \n",
       "25  A*24:02:01:01           4            donor  TCCCTCCTTACCCCATCTCA   \n",
       "26  A*24:02:01:01           4            donor  CCTTACCCCATCTCAGGGTG   \n",
       "27  A*24:02:01:01           4            donor  CTTACCCCATCTCAGGGTGA   \n",
       "28  A*24:02:01:01           4            donor  TTACCCCATCTCAGGGTGAG   \n",
       "29  A*24:02:01:01           4            donor  CTCCCTCCTTACCCCATCTC   \n",
       "30  A*24:02:01:01           4            donor  TCCCTCCTTACCCCATCTCA   \n",
       "31  A*24:02:01:01           4            donor  CCTTACCCCATCTCAGGGTG   \n",
       "32  A*24:02:01:01           4            donor  CTTACCCCATCTCAGGGTGA   \n",
       "33  A*24:02:01:01           4            donor  TTACCCCATCTCAGGGTGAG   \n",
       "34  A*24:02:01:01           5         acceptor  GTGGGCTGGGAAGATGGCTC   \n",
       "35  A*24:02:01:01           5         acceptor  TGGGCTGGGAAGATGGCTCT   \n",
       "36  A*24:02:01:01           5         acceptor  GAAGATGGCTCTGGGAAAAG   \n",
       "37  A*24:02:01:01           5         acceptor  AAGATGGCTCTGGGAAAAGA   \n",
       "38  A*24:02:01:01           5         acceptor  AGATGGCTCTGGGAAAAGAG   \n",
       "39  A*24:02:01:01           5         acceptor  GGCTCTGGGAAAAGAGGGGA   \n",
       "40  A*24:02:01:01           6         acceptor  GCTCCCTCCTTTTCTATCTG   \n",
       "41  A*24:02:01:01           6         acceptor  CTCCCTCCTTTTCTATCTGT   \n",
       "42  A*24:02:01:01           6         acceptor  TTTCTTCCCACAGATAGAAA   \n",
       "43  A*24:02:01:01           6         acceptor  CTTCCCACAGATAGAAAAGG   \n",
       "44  A*24:02:01:01           6         acceptor  TTCCCACAGATAGAAAAGGA   \n",
       "45  A*24:02:01:01           7         acceptor  CCCTGGGCACTGTCACTGCC   \n",
       "46  A*24:02:01:01           7         acceptor  CCTGGGCACTGTCACTGCCT   \n",
       "47  A*24:02:01:01           7         acceptor  CTGGGCACTGTCACTGCCTG   \n",
       "48  A*24:02:01:01           7         acceptor  CCCAGGCAGTGACAGTGCCC   \n",
       "49  A*24:02:01:01           7         acceptor  CCAGGCAGTGACAGTGCCCA   \n",
       "\n",
       "         guide seq with PAM  PAM enzyme class  CBE/ABE position  guide score  \\\n",
       "0   GACCCCGCACTCACCTGCCCAGG  NGG          CBE                14     0.115790   \n",
       "1   CGCACTCACCTGCCCAGGTCTGG  NGG          CBE                 9     0.340876   \n",
       "2   GCACTCACCTGCCCAGGTCTGGG  NGG          CBE                 8     0.473178   \n",
       "3   CACCTGCCCAGGTCTGGGTCAGG  NGG          CBE                 3     0.342684   \n",
       "4   ACCTGCCCAGGTCTGGGTCAGGG  NGG          CBE                 2     0.262362   \n",
       "5   GACCCCGCACTCACCTGCCCAGG  NGG          ABE                13     0.020324   \n",
       "6   CGCACTCACCTGCCCAGGTCTGG  NGG          ABE                 8     0.269631   \n",
       "7   GCACTCACCTGCCCAGGTCTGGG  NGG          ABE                 7     0.417329   \n",
       "8   CACCTGCCCAGGTCTGGGTCAGG  NGG          ABE                 2     0.061594   \n",
       "9   ACCTGCCCAGGTCTGGGTCAGGG  NGG          ABE                 1     0.027279   \n",
       "10  TACCTCATGGAGTGGGAGCCTGG  NGG          CBE                20     0.042578   \n",
       "11  ACCTCATGGAGTGGGAGCCTGGG  NGG          CBE                19     0.056709   \n",
       "12  CCTCATGGAGTGGGAGCCTGGGG  NGG          CBE                18     0.082248   \n",
       "13  GGAGTGGGAGCCTGGGGACGAGG  NGG          CBE                12     0.207491   \n",
       "14  GGGAGCCTGGGGACGAGGAGTGG  NGG          CBE                 7     0.670896   \n",
       "15  CCCCAGGCTCCCACTCCATGAGG  NGG          ABE                 5     0.454827   \n",
       "16  CCCGGGCCGGGGTCACTCACCGG  NGG          CBE                20     0.044239   \n",
       "17  GTCACTCACCGGCCTCGCTCTGG  NGG          CBE                 9     0.398258   \n",
       "18  CCCGGGCCGGGGTCACTCACCGG  NGG          ABE                19     0.028549   \n",
       "19  GTCACTCACCGGCCTCGCTCTGG  NGG          ABE                 8     0.343390   \n",
       "20  ATCTGGAGGGTGTGAGAACCTGG  NGG          CBE                20     0.039942   \n",
       "21  AGAACCTGGCCCCGACCCCGCGG  NGG          CBE                 6     0.860042   \n",
       "22  TGTGTCTTGGGGGGGTCTGACGG  NGG          CBE                17     0.171443   \n",
       "23  GTGTCTTGGGGGGGTCTGACGGG  NGG          CBE                16     0.194259   \n",
       "24  CTCCCTCCTTACCCCATCTCAGG  NGG          CBE                12     0.130095   \n",
       "25  TCCCTCCTTACCCCATCTCAGGG  NGG          CBE                11     0.164163   \n",
       "26  CCTTACCCCATCTCAGGGTGAGG  NGG          CBE                 6     0.789564   \n",
       "27  CTTACCCCATCTCAGGGTGAGGG  NGG          CBE                 5     0.573362   \n",
       "28  TTACCCCATCTCAGGGTGAGGGG  NGG          CBE                 4     0.392107   \n",
       "29  CTCCCTCCTTACCCCATCTCAGG  NGG          ABE                11     0.129198   \n",
       "30  TCCCTCCTTACCCCATCTCAGGG  NGG          ABE                10     0.185686   \n",
       "31  CCTTACCCCATCTCAGGGTGAGG  NGG          ABE                 5     0.604024   \n",
       "32  CTTACCCCATCTCAGGGTGAGGG  NGG          ABE                 4     0.363967   \n",
       "33  TTACCCCATCTCAGGGTGAGGGG  NGG          ABE                 3     0.188794   \n",
       "34  GTGGGCTGGGAAGATGGCTCTGG  NGG          CBE                20     0.090060   \n",
       "35  TGGGCTGGGAAGATGGCTCTGGG  NGG          CBE                19     0.118008   \n",
       "36  GAAGATGGCTCTGGGAAAAGAGG  NGG          CBE                11     0.433502   \n",
       "37  AAGATGGCTCTGGGAAAAGAGGG  NGG          CBE                10     0.494032   \n",
       "38  AGATGGCTCTGGGAAAAGAGGGG  NGG          CBE                 9     0.585956   \n",
       "39  GGCTCTGGGAAAAGAGGGGAAGG  NGG          CBE                 5     0.839644   \n",
       "40  GCTCCCTCCTTTTCTATCTGTGG  NGG          CBE                18     0.119323   \n",
       "41  CTCCCTCCTTTTCTATCTGTGGG  NGG          CBE                17     0.188146   \n",
       "42  TTTCTTCCCACAGATAGAAAAGG  NGG          ABE                12     0.035618   \n",
       "43  CTTCCCACAGATAGAAAAGGAGG  NGG          ABE                 9     0.112869   \n",
       "44  TTCCCACAGATAGAAAAGGAGGG  NGG          ABE                 8     0.187633   \n",
       "45  CCCTGGGCACTGTCACTGCCTGG  NGG          CBE                20     0.044522   \n",
       "46  CCTGGGCACTGTCACTGCCTGGG  NGG          CBE                19     0.059258   \n",
       "47  CTGGGCACTGTCACTGCCTGGGG  NGG          CBE                18     0.085840   \n",
       "48  CCCAGGCAGTGACAGTGCCCAGG  NGG          ABE                 4     0.238371   \n",
       "49  CCAGGCAGTGACAGTGCCCAGGG  NGG          ABE                 3     0.112914   \n",
       "\n",
       "    cDNA distuption pct  \n",
       "0              0.067395  \n",
       "1              0.067395  \n",
       "2              0.067395  \n",
       "3              0.067395  \n",
       "4              0.067395  \n",
       "5              0.067395  \n",
       "6              0.067395  \n",
       "7              0.067395  \n",
       "8              0.067395  \n",
       "9              0.067395  \n",
       "10             0.313297  \n",
       "11             0.313297  \n",
       "12             0.313297  \n",
       "13             0.313297  \n",
       "14             0.313297  \n",
       "15             0.313297  \n",
       "16             0.313297  \n",
       "17             0.313297  \n",
       "18             0.313297  \n",
       "19             0.313297  \n",
       "20             0.564663  \n",
       "21             0.564663  \n",
       "22             0.816029  \n",
       "23             0.816029  \n",
       "24             0.816029  \n",
       "25             0.816029  \n",
       "26             0.816029  \n",
       "27             0.816029  \n",
       "28             0.816029  \n",
       "29             0.816029  \n",
       "30             0.816029  \n",
       "31             0.816029  \n",
       "32             0.816029  \n",
       "33             0.816029  \n",
       "34             0.922587  \n",
       "35             0.922587  \n",
       "36             0.922587  \n",
       "37             0.922587  \n",
       "38             0.922587  \n",
       "39             0.922587  \n",
       "40             0.952641  \n",
       "41             0.952641  \n",
       "42             0.952641  \n",
       "43             0.952641  \n",
       "44             0.952641  \n",
       "45             0.996357  \n",
       "46             0.996357  \n",
       "47             0.996357  \n",
       "48             0.996357  \n",
       "49             0.996357  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for HLA project: \n",
    "motif_weights_filepath = \"../motif_weights.tsv\"\n",
    "position_weights_filepath = \"../position_weights.tsv\"\n",
    "motif_weights_dict, position_weights_dict = parse_motif_position_weights(motif_weights_filepath, position_weights_filepath)\n",
    "\n",
    "A_24_02_info = parse_json(\"HLA00050.json\")\n",
    "genomic_features = A_24_02_info['genomic_features']\n",
    "coding_features = A_24_02_info['coding_features']\n",
    "genomic_sequence = A_24_02_info['genomic_sequence']\n",
    "coding_sequence = A_24_02_info['sequences']['coding']\n",
    "\n",
    "test1 = spliceR_main(\n",
    "    genomic_features,\n",
    "    coding_features,\n",
    "    genomic_sequence,\n",
    "    coding_sequence,\n",
    "    motif_weights_dict,\n",
    "    position_weights_dict,\n",
    "    output_as_df = True,\n",
    "    haplotype_name = \"A*24:02:01:01\",\n",
    "    guide_length = 20, \n",
    "    pam = \"NGG\",\n",
    "    flank_length = 20,\n",
    "    CBE = True, \n",
    "    ABE = True,\n",
    "    SA = True, \n",
    "    SD = True,\n",
    "    logistic_adjust = 1\n",
    ")\n",
    "test1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'haplotype_name': 'A*01:02:01:01',\n",
       " 'prevalence': ('C', 'C'),\n",
       " 'genomic_features': [{'length': 300, 'start': 1, 'type': '5utr'},\n",
       "  {'length': 73,\n",
       "   'number': '1',\n",
       "   'partial': False,\n",
       "   'start': 301,\n",
       "   'type': 'exon'},\n",
       "  {'length': 130,\n",
       "   'number': '1',\n",
       "   'partial': False,\n",
       "   'start': 374,\n",
       "   'type': 'intron'},\n",
       "  {'length': 270,\n",
       "   'number': '2',\n",
       "   'partial': False,\n",
       "   'start': 504,\n",
       "   'type': 'exon'},\n",
       "  {'length': 241,\n",
       "   'number': '2',\n",
       "   'partial': False,\n",
       "   'start': 774,\n",
       "   'type': 'intron'},\n",
       "  {'length': 276,\n",
       "   'number': '3',\n",
       "   'partial': False,\n",
       "   'start': 1015,\n",
       "   'type': 'exon'},\n",
       "  {'length': 579,\n",
       "   'number': '3',\n",
       "   'partial': False,\n",
       "   'start': 1291,\n",
       "   'type': 'intron'},\n",
       "  {'length': 276,\n",
       "   'number': '4',\n",
       "   'partial': False,\n",
       "   'start': 1870,\n",
       "   'type': 'exon'},\n",
       "  {'length': 102,\n",
       "   'number': '4',\n",
       "   'partial': False,\n",
       "   'start': 2146,\n",
       "   'type': 'intron'},\n",
       "  {'length': 117,\n",
       "   'number': '5',\n",
       "   'partial': False,\n",
       "   'start': 2248,\n",
       "   'type': 'exon'},\n",
       "  {'length': 442,\n",
       "   'number': '5',\n",
       "   'partial': False,\n",
       "   'start': 2365,\n",
       "   'type': 'intron'},\n",
       "  {'length': 33,\n",
       "   'number': '6',\n",
       "   'partial': False,\n",
       "   'start': 2807,\n",
       "   'type': 'exon'},\n",
       "  {'length': 142,\n",
       "   'number': '6',\n",
       "   'partial': False,\n",
       "   'start': 2840,\n",
       "   'type': 'intron'},\n",
       "  {'length': 48,\n",
       "   'number': '7',\n",
       "   'partial': False,\n",
       "   'start': 2982,\n",
       "   'type': 'exon'},\n",
       "  {'length': 169,\n",
       "   'number': '7',\n",
       "   'partial': False,\n",
       "   'start': 3030,\n",
       "   'type': 'intron'},\n",
       "  {'length': 5,\n",
       "   'number': '8',\n",
       "   'partial': False,\n",
       "   'start': 3199,\n",
       "   'type': 'exon'},\n",
       "  {'length': 300, 'start': 3204, 'type': '3utr'}],\n",
       " 'coding_features': [{'length': 73,\n",
       "   'number': '1',\n",
       "   'partial': False,\n",
       "   'start': 1,\n",
       "   'type': 'exon'},\n",
       "  {'length': 270,\n",
       "   'number': '2',\n",
       "   'partial': False,\n",
       "   'start': 74,\n",
       "   'type': 'exon'},\n",
       "  {'length': 276,\n",
       "   'number': '3',\n",
       "   'partial': False,\n",
       "   'start': 344,\n",
       "   'type': 'exon'},\n",
       "  {'length': 276,\n",
       "   'number': '4',\n",
       "   'partial': False,\n",
       "   'start': 620,\n",
       "   'type': 'exon'},\n",
       "  {'length': 117,\n",
       "   'number': '5',\n",
       "   'partial': False,\n",
       "   'start': 896,\n",
       "   'type': 'exon'},\n",
       "  {'length': 33,\n",
       "   'number': '6',\n",
       "   'partial': False,\n",
       "   'start': 1013,\n",
       "   'type': 'exon'},\n",
       "  {'length': 48,\n",
       "   'number': '7',\n",
       "   'partial': False,\n",
       "   'start': 1046,\n",
       "   'type': 'exon'},\n",
       "  {'length': 5,\n",
       "   'number': '8',\n",
       "   'partial': False,\n",
       "   'start': 1094,\n",
       "   'type': 'exon'}],\n",
       " 'sequences': {'coding': 'ATGGCCGTCATGGCGCCCCGAACCCTCCTCCTGCTACTCTCGGGGGCCCTGGCCCTGACCCAGACCTGGGCGGGCTCCCACTCCATGAGGTATTTCTCCACATCCGTGTCCCGGCCCGGCAGTGGAGAGCCCCGCTTCATCGCAGTGGGCTACGTGGACGACACGCAGTTCGTGCGGTTCGACAGCGACGCCGCGAGCCAGAAGATGGAGCCGCGGGCGCCGTGGATAGAGCAGGAGGGGCCGGAGTATTGGGACCAGGAGACACGGAATATGAAGGCCCACTCACAGACTGACCGAGCGAACCTGGGGACCCTGCGCGGCTACTACAACCAGAGCGAGGACGGTTCTCACACCATCCAGATAATGTATGGCTGCGACGTGGGGCCGGACGGGCGCTTCCTCCGCGGGTACCGGCAGGACGCCTACGACGGCAAGGATTACATCGCCCTGAACGAGGACCTGCGCTCTTGGACCGCGGCGGACATGGCAGCTCAGATCACCAAGCGCAAGTGGGAGGCGGTCCATGCGGCGGAGCAGCGGAGAGTCTACCTGGAGGGCCGGTGCGTGGACGGGCTCCGCAGATACCTGGAGAACGGGAAGGAGACGCTGCAGCGCACGGACCCCCCCAAGACACATATGACCCACCACCCCATCTCTGACCATGAGGCCACCCTGAGGTGCTGGGCCCTGGGCTTCTACCCTGCGGAGATCACACTGACCTGGCAGCGGGATGGGGAGGACCAGACCCAGGACACGGAGCTCGTGGAGACCAGGCCTGCAGGGGATGGAACCTTCCAGAAGTGGGCGGCTGTGGTGGTGCCTTCTGGAGAGGAGCAGAGATACACCTGCCATGTGCAGCATGAGGGTCTGCCCAAGCCCCTCACCCTGAGATGGGAGCTGTCTTCCCAGCCCACCATCCCCATCGTGGGCATCATTGCTGGCCTGGTTCTCCTTGGAGCTGTGATCACTGGAGCTGTGGTCGCTGCCGTGATGTGGAGGAGGAAGAGCTCAGATAGAAAAGGAGGGAGTTACACTCAGGCTGCAAGCAGTGACAGTGCCCAGGGCTCTGATGTGTCTCTCACAGCTTGTAAAGTGTGA',\n",
       "  'codon_start': 1,\n",
       "  'genomic': 'CAGGAGCAGAGGGGTCAGGGCGAAGTCCCAGGGCCCCAGGCGTGGCTCTCAGGGTCTCAGGCCCCGAAGGCGGTGTATGGATTGGGGAGTCCCAGCCTTGGGGATTCCCCAACTCCGCAGTTTCTTTTCTCCCTCTCCCAACCTACGTAGGGTCCTTCATCCTGGATACTCACGACGCGGACCCAGTTCTCACTCCCATTGGGTGTCGGGTTTCCAGAGAAGCCAATCAGTGTCGTCGCGGTCGCTGTTCTAAAGTCCGCACGCACCCACCGGGACTCAGATTCTCCCCAGACGCCGAGGATGGCCGTCATGGCGCCCCGAACCCTCCTCCTGCTACTCTCGGGGGCCCTGGCCCTGACCCAGACCTGGGCGGGTGAGTGCGGGGTCGGGAGGGAAACCGCCTCTGCGGGGAGAAGCAAGGGGCCCTCCTGGCGGGGGCGCAGGACCGGGGGAGCCGCGCCGGGAGGAGGGTCGGTCAGGTCTCAGCCACTGCTCGCCCCCAGGCTCCCACTCCATGAGGTATTTCTCCACATCCGTGTCCCGGCCCGGCAGTGGAGAGCCCCGCTTCATCGCAGTGGGCTACGTGGACGACACGCAGTTCGTGCGGTTCGACAGCGACGCCGCGAGCCAGAAGATGGAGCCGCGGGCGCCGTGGATAGAGCAGGAGGGGCCGGAGTATTGGGACCAGGAGACACGGAATATGAAGGCCCACTCACAGACTGACCGAGCGAACCTGGGGACCCTGCGCGGCTACTACAACCAGAGCGAGGACGGTGAGTGACCCCGGCCCGGGGCGCAGGTCACGACCCCTCATCCCCCACGGACGGGCCAGGTCGCCCACAGTCTCCGGGTCCGAGATCCACCCCGAAGCCGCGGGACTCCGAGACCCTTGTCCCGGGAGAGGCCCAGGCGCCTTTACCCGGTTTCATTTTCAGTTTAGGCCAAAAATCCCCCCGGGTTGGTCGGGGCGGGGCGGGGCTCGGGGGACTGGGCTGACCGCGGGGTCGGGGCCAGGTTCTCACACCATCCAGATAATGTATGGCTGCGACGTGGGGCCGGACGGGCGCTTCCTCCGCGGGTACCGGCAGGACGCCTACGACGGCAAGGATTACATCGCCCTGAACGAGGACCTGCGCTCTTGGACCGCGGCGGACATGGCAGCTCAGATCACCAAGCGCAAGTGGGAGGCGGTCCATGCGGCGGAGCAGCGGAGAGTCTACCTGGAGGGCCGGTGCGTGGACGGGCTCCGCAGATACCTGGAGAACGGGAAGGAGACGCTGCAGCGCACGGGTACCAGGGGCCACGGGGCGCCTCCCTGATCGCCTATAGATCTCCCGGGCTGGCCTCCCACAAGGAGGGGAGACAATTGGGACCAACACTAGAATATCACCCTCCCTCTGGTCCTGAGGGAGAGGAATCCTCCTGGGTTTCCAGATCCTGTACCAGAGAGTGACTCTGAGGTTCCGCCCTGCTCTCTGACACAATTAAGGGATAAAATCTCTGAAGGAGTGACGGGAAGACGATCCCTCGAATACTGATGAGTGGTTCCCTTTGACACCGGCAGCAGCCTTGGGCCCGTGACTTTTCCTCTCAGGCCTTGTTCTCTGCTTCACACTCAATGTGTGTGGGGGTCTGAGTCCAGCACTTCTGAGTCTCTCAGCCTCCACTCAGGTCAGGACCAGAAGTCGCTGTTCCCTTCTCAGGGAATAGAAGATTATCCCAGGTGCCTGTGTCCAGGCTGGTGTCTGGGTTCTGTGCTCTCTTCCCCATCCCGGGTGTCCTGTCCATTCTCAAGATGGCCACATGCGTGCTGGTGGAGTGTCCCATGACAGATGCAAAATGCCTGAATTTTCTGACTCTTCCCGTCAGACCCCCCCAAGACACATATGACCCACCACCCCATCTCTGACCATGAGGCCACCCTGAGGTGCTGGGCCCTGGGCTTCTACCCTGCGGAGATCACACTGACCTGGCAGCGGGATGGGGAGGACCAGACCCAGGACACGGAGCTCGTGGAGACCAGGCCTGCAGGGGATGGAACCTTCCAGAAGTGGGCGGCTGTGGTGGTGCCTTCTGGAGAGGAGCAGAGATACACCTGCCATGTGCAGCATGAGGGTCTGCCCAAGCCCCTCACCCTGAGATGGGGTAAGGAGGGAGATGGGGGTGTCATGTCTCTTAGGGAAAGCAGGAGCCTCTCTGGAGACCTTTAGCAGGGTCAGGGCCCCTCACCTTCCCCTCTTTTCCCAGAGCTGTCTTCCCAGCCCACCATCCCCATCGTGGGCATCATTGCTGGCCTGGTTCTCCTTGGAGCTGTGATCACTGGAGCTGTGGTCGCTGCCGTGATGTGGAGGAGGAAGAGCTCAGGTGGAGAAGGGGTGAAGGGTGGGGTCTGAGATTTCTTGTCTCACTGAGGGTTCCAAGCCCCAGCTAGAAATGTGCCCTGTCTCATTACTGGGAAGCACCTTCCACAATCATGGGCCGACCCAGCCTGGGCCCTGTGTGCCAGCACTTACTCTTTTGTAAAGCACCTGTTAAAATGAAGGACAGATTTATCACCTTGATTACGGCGGTGATGGGACCTGATCCCAGCAGTCACAAGTCACAGGGGAAGGTCCCTGAGGACAGACCTCAGGAGGGCTATTGGTCCAGGACCCACACCTGCTTTCTTCATGTTTCCTGATCCCGCCCTGGGTCTGCAGTCACACATTTCTGGAAACTTCTCTGGGGTCCAAGACTAGGAGGTTCCTCTAGGACCTTAAGGCCCTGGCTCCTTTCTGGTATCTCACAGGACATTTTCTTCCCACAGATAGAAAAGGAGGGAGTTACACTCAGGCTGCAAGTAAGTATGAAGGAGGCTGATGCCTGAGGTCCTTGGGATATTGTGTTTGGGAGCCCATGGGGGAGCTCACCCACCCCACAATTCCTCCTCTAGCCACATCTTCTGTGGGATCTGACCAGGTTCTGTTTTTGTTCTACCCCAGGCAGTGACAGTGCCCAGGGCTCTGATGTGTCTCTCACAGCTTGTAAAGGTGAGAGCTTGGAGGGCCTGATGTGTGTTGGGTGTTGGGTGGAACAGTGGACACAGCTGTGCTATGGGGTTTCTTTGCGTTGGATGTATTGAGCATGCGATGGGCTGTTTAAGGTGTGACCCCTCACTGTGATGGATATGAATTTGTTCATGAATATTTTTTTCTATAGTGTGAGACAGCTGCCTTGTGTGGGACTGAGAGGCAAGAGTTGTTCCTGCCCTTCCCTTTGTGACTTGAAGAACCCTGACTTTGTTTCTGCAAAGGCACCTGCATGTGTCTGTGTTCGTGTAGGCATAATGTGAGGAGGTGGGGAGAGCACCCCACCCCCATGTCCACCATGACCCTCTTCCCACGCTGACCTGTGCTCCCTCTCCAATCATCTTTCCTGTTCCAGAGAGGTGGGGCTGAGGTGTCTCCATCTCTGTCTCAACTTCATGGTGCACTGAGCTGTAACTTCTTCCTTCCCTATTAAAA',\n",
       "  'protein': 'MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFSTSVSRPGSGEPRFIAVGYVDDTQFVRFDSDAASQKMEPRAPWIEQEGPEYWDQETRNMKAHSQTDRANLGTLRGYYNQSEDGSHTIQIMYGCDVGPDGRFLRGYRQDAYDGKDYIALNEDLRSWTAADMAAQITKRKWEAVHAAEQRRVYLEGRCVDGLRRYLENGKETLQRTDPPKTHMTHHPISDHEATLRCWALGFYPAEITLTWQRDGEDQTQDTELVETRPAGDGTFQKWAAVVVPSGEEQRYTCHVQHEGLPKPLTLRWELSSQPTIPIVGIIAGLVLLGAVITGAVVAAVMWRRKSSDRKGGSYTQAASSDSAQGSDVSLTACKV'},\n",
       " 'genomic_sequence': 'CAGGAGCAGAGGGGTCAGGGCGAAGTCCCAGGGCCCCAGGCGTGGCTCTCAGGGTCTCAGGCCCCGAAGGCGGTGTATGGATTGGGGAGTCCCAGCCTTGGGGATTCCCCAACTCCGCAGTTTCTTTTCTCCCTCTCCCAACCTACGTAGGGTCCTTCATCCTGGATACTCACGACGCGGACCCAGTTCTCACTCCCATTGGGTGTCGGGTTTCCAGAGAAGCCAATCAGTGTCGTCGCGGTCGCTGTTCTAAAGTCCGCACGCACCCACCGGGACTCAGATTCTCCCCAGACGCCGAGGATGGCCGTCATGGCGCCCCGAACCCTCCTCCTGCTACTCTCGGGGGCCCTGGCCCTGACCCAGACCTGGGCGGGTGAGTGCGGGGTCGGGAGGGAAACCGCCTCTGCGGGGAGAAGCAAGGGGCCCTCCTGGCGGGGGCGCAGGACCGGGGGAGCCGCGCCGGGAGGAGGGTCGGTCAGGTCTCAGCCACTGCTCGCCCCCAGGCTCCCACTCCATGAGGTATTTCTCCACATCCGTGTCCCGGCCCGGCAGTGGAGAGCCCCGCTTCATCGCAGTGGGCTACGTGGACGACACGCAGTTCGTGCGGTTCGACAGCGACGCCGCGAGCCAGAAGATGGAGCCGCGGGCGCCGTGGATAGAGCAGGAGGGGCCGGAGTATTGGGACCAGGAGACACGGAATATGAAGGCCCACTCACAGACTGACCGAGCGAACCTGGGGACCCTGCGCGGCTACTACAACCAGAGCGAGGACGGTGAGTGACCCCGGCCCGGGGCGCAGGTCACGACCCCTCATCCCCCACGGACGGGCCAGGTCGCCCACAGTCTCCGGGTCCGAGATCCACCCCGAAGCCGCGGGACTCCGAGACCCTTGTCCCGGGAGAGGCCCAGGCGCCTTTACCCGGTTTCATTTTCAGTTTAGGCCAAAAATCCCCCCGGGTTGGTCGGGGCGGGGCGGGGCTCGGGGGACTGGGCTGACCGCGGGGTCGGGGCCAGGTTCTCACACCATCCAGATAATGTATGGCTGCGACGTGGGGCCGGACGGGCGCTTCCTCCGCGGGTACCGGCAGGACGCCTACGACGGCAAGGATTACATCGCCCTGAACGAGGACCTGCGCTCTTGGACCGCGGCGGACATGGCAGCTCAGATCACCAAGCGCAAGTGGGAGGCGGTCCATGCGGCGGAGCAGCGGAGAGTCTACCTGGAGGGCCGGTGCGTGGACGGGCTCCGCAGATACCTGGAGAACGGGAAGGAGACGCTGCAGCGCACGGGTACCAGGGGCCACGGGGCGCCTCCCTGATCGCCTATAGATCTCCCGGGCTGGCCTCCCACAAGGAGGGGAGACAATTGGGACCAACACTAGAATATCACCCTCCCTCTGGTCCTGAGGGAGAGGAATCCTCCTGGGTTTCCAGATCCTGTACCAGAGAGTGACTCTGAGGTTCCGCCCTGCTCTCTGACACAATTAAGGGATAAAATCTCTGAAGGAGTGACGGGAAGACGATCCCTCGAATACTGATGAGTGGTTCCCTTTGACACCGGCAGCAGCCTTGGGCCCGTGACTTTTCCTCTCAGGCCTTGTTCTCTGCTTCACACTCAATGTGTGTGGGGGTCTGAGTCCAGCACTTCTGAGTCTCTCAGCCTCCACTCAGGTCAGGACCAGAAGTCGCTGTTCCCTTCTCAGGGAATAGAAGATTATCCCAGGTGCCTGTGTCCAGGCTGGTGTCTGGGTTCTGTGCTCTCTTCCCCATCCCGGGTGTCCTGTCCATTCTCAAGATGGCCACATGCGTGCTGGTGGAGTGTCCCATGACAGATGCAAAATGCCTGAATTTTCTGACTCTTCCCGTCAGACCCCCCCAAGACACATATGACCCACCACCCCATCTCTGACCATGAGGCCACCCTGAGGTGCTGGGCCCTGGGCTTCTACCCTGCGGAGATCACACTGACCTGGCAGCGGGATGGGGAGGACCAGACCCAGGACACGGAGCTCGTGGAGACCAGGCCTGCAGGGGATGGAACCTTCCAGAAGTGGGCGGCTGTGGTGGTGCCTTCTGGAGAGGAGCAGAGATACACCTGCCATGTGCAGCATGAGGGTCTGCCCAAGCCCCTCACCCTGAGATGGGGTAAGGAGGGAGATGGGGGTGTCATGTCTCTTAGGGAAAGCAGGAGCCTCTCTGGAGACCTTTAGCAGGGTCAGGGCCCCTCACCTTCCCCTCTTTTCCCAGAGCTGTCTTCCCAGCCCACCATCCCCATCGTGGGCATCATTGCTGGCCTGGTTCTCCTTGGAGCTGTGATCACTGGAGCTGTGGTCGCTGCCGTGATGTGGAGGAGGAAGAGCTCAGGTGGAGAAGGGGTGAAGGGTGGGGTCTGAGATTTCTTGTCTCACTGAGGGTTCCAAGCCCCAGCTAGAAATGTGCCCTGTCTCATTACTGGGAAGCACCTTCCACAATCATGGGCCGACCCAGCCTGGGCCCTGTGTGCCAGCACTTACTCTTTTGTAAAGCACCTGTTAAAATGAAGGACAGATTTATCACCTTGATTACGGCGGTGATGGGACCTGATCCCAGCAGTCACAAGTCACAGGGGAAGGTCCCTGAGGACAGACCTCAGGAGGGCTATTGGTCCAGGACCCACACCTGCTTTCTTCATGTTTCCTGATCCCGCCCTGGGTCTGCAGTCACACATTTCTGGAAACTTCTCTGGGGTCCAAGACTAGGAGGTTCCTCTAGGACCTTAAGGCCCTGGCTCCTTTCTGGTATCTCACAGGACATTTTCTTCCCACAGATAGAAAAGGAGGGAGTTACACTCAGGCTGCAAGTAAGTATGAAGGAGGCTGATGCCTGAGGTCCTTGGGATATTGTGTTTGGGAGCCCATGGGGGAGCTCACCCACCCCACAATTCCTCCTCTAGCCACATCTTCTGTGGGATCTGACCAGGTTCTGTTTTTGTTCTACCCCAGGCAGTGACAGTGCCCAGGGCTCTGATGTGTCTCTCACAGCTTGTAAAGGTGAGAGCTTGGAGGGCCTGATGTGTGTTGGGTGTTGGGTGGAACAGTGGACACAGCTGTGCTATGGGGTTTCTTTGCGTTGGATGTATTGAGCATGCGATGGGCTGTTTAAGGTGTGACCCCTCACTGTGATGGATATGAATTTGTTCATGAATATTTTTTTCTATAGTGTGAGACAGCTGCCTTGTGTGGGACTGAGAGGCAAGAGTTGTTCCTGCCCTTCCCTTTGTGACTTGAAGAACCCTGACTTTGTTTCTGCAAAGGCACCTGCATGTGTCTGTGTTCGTGTAGGCATAATGTGAGGAGGTGGGGAGAGCACCCCACCCCCATGTCCACCATGACCCTCTTCCCACGCTGACCTGTGCTCCCTCTCCAATCATCTTTCCTGTTCCAGAGAGGTGGGGCTGAGGTGTCTCCATCTCTGTCTCAACTTCATGGTGCACTGAGCTGTAACTTCTTCCTTCCCTATTAAAA'}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "A_24_02_info_frm_url = parse_json(file_url=\"https://www.ebi.ac.uk/cgi-bin/ipd/api/allele/HLA00002\")\n",
    "A_24_02_info_frm_url"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "HLAoct2025",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
